From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dejavu <thedeejayv1@gmail.com>
Date: Sun, 8 Mar 2026 15:30:08 +0200
Subject: [PATCH] Rename package segment to starforge


diff --git a/src/main/java/dev/orhidea/entropy/entity/EntropyFallingBlock.java b/src/main/java/dev/orhidea/starforge/entity/EntropyFallingBlock.java
similarity index 86%
rename from src/main/java/dev/orhidea/entropy/entity/EntropyFallingBlock.java
rename to src/main/java/dev/orhidea/starforge/entity/EntropyFallingBlock.java
index c2b42342d46eb591725c2cf88edc780172bc2013..cac8314d4bed584fbb96e303ba9c96a13cd0d7d3 100644
--- a/src/main/java/dev/orhidea/entropy/entity/EntropyFallingBlock.java
+++ b/src/main/java/dev/orhidea/starforge/entity/EntropyFallingBlock.java
@@ -1,4 +1,4 @@
-package dev.orhidea.entropy.entity;
+package dev.orhidea.starforge.entity;
 
 import org.bukkit.entity.FallingBlock;
 
@@ -11,3 +11,4 @@ public interface EntropyFallingBlock extends FallingBlock {
     public int getCount();
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/entity/EntropyTntPrimed.java b/src/main/java/dev/orhidea/starforge/entity/EntropyTntPrimed.java
similarity index 87%
rename from src/main/java/dev/orhidea/entropy/entity/EntropyTntPrimed.java
rename to src/main/java/dev/orhidea/starforge/entity/EntropyTntPrimed.java
index 64f0b50b083138fbc643b517a4075cbe8e5d9259..823afa4353d9dfc035acd85d04f0305c602e0e64 100644
--- a/src/main/java/dev/orhidea/entropy/entity/EntropyTntPrimed.java
+++ b/src/main/java/dev/orhidea/starforge/entity/EntropyTntPrimed.java
@@ -1,4 +1,4 @@
-package dev.orhidea.entropy.entity;
+package dev.orhidea.starforge.entity;
 
 import org.bukkit.entity.TNTPrimed;
 
@@ -15,3 +15,4 @@ public interface EntropyTntPrimed extends TNTPrimed {
 }
 
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/AssignDurabilityEvent.java b/src/main/java/dev/orhidea/starforge/event/AssignDurabilityEvent.java
similarity index 96%
rename from src/main/java/dev/orhidea/entropy/event/AssignDurabilityEvent.java
rename to src/main/java/dev/orhidea/starforge/event/AssignDurabilityEvent.java
index 649a32f0613c31e6c0e3b88eacb20c25b0929a47..d09229cd4c27a543cd7be558b567f30ec0fc4013 100644
--- a/src/main/java/dev/orhidea/entropy/event/AssignDurabilityEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/AssignDurabilityEvent.java
@@ -1,4 +1,4 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.Material;
@@ -44,3 +44,4 @@ public class AssignDurabilityEvent extends Event {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/BubbleColumnEvent.java b/src/main/java/dev/orhidea/starforge/event/BubbleColumnEvent.java
similarity index 83%
rename from src/main/java/dev/orhidea/entropy/event/BubbleColumnEvent.java
rename to src/main/java/dev/orhidea/starforge/event/BubbleColumnEvent.java
index 5bffb577af3de58436ceaf3d6c1dac8c9785e27e..a2bece6f2b972591dc88a8c750655e62e318cd97 100644
--- a/src/main/java/dev/orhidea/entropy/event/BubbleColumnEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/BubbleColumnEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
-import dev.orhidea.entropy.event.base.EntropyBooleanEvent;
+import dev.orhidea.starforge.event.base.EntropyBooleanEvent;
 
 public class BubbleColumnEvent extends EntropyBooleanEvent {
     private final Entity entity;
@@ -22,3 +22,4 @@ public class BubbleColumnEvent extends EntropyBooleanEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/CombinedExplosionEvent.java b/src/main/java/dev/orhidea/starforge/event/CombinedExplosionEvent.java
similarity index 75%
rename from src/main/java/dev/orhidea/entropy/event/CombinedExplosionEvent.java
rename to src/main/java/dev/orhidea/starforge/event/CombinedExplosionEvent.java
index f2d8225630ae24e98888d32d6e33b855ea67e676..d2d0306a5a501edc5df8197ae7073a899ea7cd53 100644
--- a/src/main/java/dev/orhidea/entropy/event/CombinedExplosionEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/CombinedExplosionEvent.java
@@ -1,9 +1,9 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.event.base.EntropyExplosionEvent;
+import dev.orhidea.starforge.event.base.EntropyExplosionEvent;
 
 public class CombinedExplosionEvent extends EntropyExplosionEvent {
     public CombinedExplosionEvent(boolean value, Entity source, @NotNull Location location) {
@@ -11,3 +11,4 @@ public class CombinedExplosionEvent extends EntropyExplosionEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/ConcreteHardenEvent.java b/src/main/java/dev/orhidea/starforge/event/ConcreteHardenEvent.java
similarity index 60%
rename from src/main/java/dev/orhidea/entropy/event/ConcreteHardenEvent.java
rename to src/main/java/dev/orhidea/starforge/event/ConcreteHardenEvent.java
index ac90910e22876e1bf854530909aa5ec558511b19..a52c779b880f5d676015a9eb1d5e7e2ac0c40947 100644
--- a/src/main/java/dev/orhidea/entropy/event/ConcreteHardenEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/ConcreteHardenEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.entity.EntropyFallingBlock;
-import dev.orhidea.entropy.event.base.EntropyFallingBlockEvent;
+import dev.orhidea.starforge.entity.EntropyFallingBlock;
+import dev.orhidea.starforge.event.base.EntropyFallingBlockEvent;
 
 public class ConcreteHardenEvent extends EntropyFallingBlockEvent {
     public ConcreteHardenEvent(boolean value, @NotNull EntropyFallingBlock fallingBlock) {
@@ -10,3 +10,4 @@ public class ConcreteHardenEvent extends EntropyFallingBlockEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/DetermineFallingBlockExposureEvent.java b/src/main/java/dev/orhidea/starforge/event/DetermineFallingBlockExposureEvent.java
similarity index 77%
rename from src/main/java/dev/orhidea/entropy/event/DetermineFallingBlockExposureEvent.java
rename to src/main/java/dev/orhidea/starforge/event/DetermineFallingBlockExposureEvent.java
index 3109333db0674e92d9b5090a4d64a420f31bc849..d5b6cfa4308900ef8e01cc09c950483e75bc017a 100644
--- a/src/main/java/dev/orhidea/entropy/event/DetermineFallingBlockExposureEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/DetermineFallingBlockExposureEvent.java
@@ -1,9 +1,9 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.event.base.EntropyExplosionEvent;
+import dev.orhidea.starforge.event.base.EntropyExplosionEvent;
 
 public class DetermineFallingBlockExposureEvent extends EntropyExplosionEvent {
     public DetermineFallingBlockExposureEvent(boolean value, Entity source, @NotNull Location location) {
@@ -11,3 +11,4 @@ public class DetermineFallingBlockExposureEvent extends EntropyExplosionEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/EntityMergeEvent.java b/src/main/java/dev/orhidea/starforge/event/EntityMergeEvent.java
similarity index 86%
rename from src/main/java/dev/orhidea/entropy/event/EntityMergeEvent.java
rename to src/main/java/dev/orhidea/starforge/event/EntityMergeEvent.java
index 56c809f3d5db43ebda49c38ddb00233eb8ae39cb..4977b1f8b92c3b97b3d84c3cd5441dbb6f9f49f6 100644
--- a/src/main/java/dev/orhidea/entropy/event/EntityMergeEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/EntityMergeEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.event.base.EntropyBooleanEvent;
+import dev.orhidea.starforge.event.base.EntropyBooleanEvent;
 
 public class EntityMergeEvent extends EntropyBooleanEvent {
     private final Entity mergeIntoEntity;
@@ -23,3 +23,4 @@ public class EntityMergeEvent extends EntropyBooleanEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/ExplodeLavaEvent.java b/src/main/java/dev/orhidea/starforge/event/ExplodeLavaEvent.java
similarity index 75%
rename from src/main/java/dev/orhidea/entropy/event/ExplodeLavaEvent.java
rename to src/main/java/dev/orhidea/starforge/event/ExplodeLavaEvent.java
index eaf3926c9071cc0461c07b17f45a948cd59e36c3..8bdf3205b47af6f28d31a9d6c0a641835ca41145 100644
--- a/src/main/java/dev/orhidea/entropy/event/ExplodeLavaEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/ExplodeLavaEvent.java
@@ -1,9 +1,9 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.event.base.EntropyExplosionEvent;
+import dev.orhidea.starforge.event.base.EntropyExplosionEvent;
 
 public class ExplodeLavaEvent extends EntropyExplosionEvent {
     public ExplodeLavaEvent(boolean value, Entity source, @NotNull Location location) {
@@ -11,3 +11,4 @@ public class ExplodeLavaEvent extends EntropyExplosionEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/ExplodeWaterloggedEvent.java b/src/main/java/dev/orhidea/starforge/event/ExplodeWaterloggedEvent.java
similarity index 75%
rename from src/main/java/dev/orhidea/entropy/event/ExplodeWaterloggedEvent.java
rename to src/main/java/dev/orhidea/starforge/event/ExplodeWaterloggedEvent.java
index 4c51237407be38cdbeba34c02349c7e6a8c6edff..cadf4ff0598bf6e3814a80ac3a368fb5eccac97a 100644
--- a/src/main/java/dev/orhidea/entropy/event/ExplodeWaterloggedEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/ExplodeWaterloggedEvent.java
@@ -1,9 +1,9 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.event.base.EntropyExplosionEvent;
+import dev.orhidea.starforge.event.base.EntropyExplosionEvent;
 
 public class ExplodeWaterloggedEvent extends EntropyExplosionEvent {
     public ExplodeWaterloggedEvent(boolean value, Entity source, @NotNull Location location) {
@@ -11,3 +11,4 @@ public class ExplodeWaterloggedEvent extends EntropyExplosionEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/HoneyBlockEvent.java b/src/main/java/dev/orhidea/starforge/event/HoneyBlockEvent.java
similarity index 83%
rename from src/main/java/dev/orhidea/entropy/event/HoneyBlockEvent.java
rename to src/main/java/dev/orhidea/starforge/event/HoneyBlockEvent.java
index 30f56bf364b52557a8c6657d348059701710d9ad..e0a6b9d4f2f805fb0d5f73f5ec039b777b5d6f34 100644
--- a/src/main/java/dev/orhidea/entropy/event/HoneyBlockEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/HoneyBlockEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
-import dev.orhidea.entropy.event.base.EntropyBooleanEvent;
+import dev.orhidea.starforge.event.base.EntropyBooleanEvent;
 
 public class HoneyBlockEvent extends EntropyBooleanEvent {
     private final Entity entity;
@@ -22,3 +22,4 @@ public class HoneyBlockEvent extends EntropyBooleanEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/IsInCannonEvent.java b/src/main/java/dev/orhidea/starforge/event/IsInCannonEvent.java
similarity index 83%
rename from src/main/java/dev/orhidea/entropy/event/IsInCannonEvent.java
rename to src/main/java/dev/orhidea/starforge/event/IsInCannonEvent.java
index 71e92583c402db848a90bb72f3a5c4da0ee0e761..2742753059a3089dd2ee6a4b993a17eccc06961b 100644
--- a/src/main/java/dev/orhidea/entropy/event/IsInCannonEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/IsInCannonEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
-import dev.orhidea.entropy.event.base.EntropyBooleanEvent;
+import dev.orhidea.starforge.event.base.EntropyBooleanEvent;
 
 public class IsInCannonEvent extends EntropyBooleanEvent {
     private final Entity entity;
@@ -22,3 +22,4 @@ public class IsInCannonEvent extends EntropyBooleanEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/NonFullBlockCollideEvent.java b/src/main/java/dev/orhidea/starforge/event/NonFullBlockCollideEvent.java
similarity index 83%
rename from src/main/java/dev/orhidea/entropy/event/NonFullBlockCollideEvent.java
rename to src/main/java/dev/orhidea/starforge/event/NonFullBlockCollideEvent.java
index 8b5b42bef6650335480d26b10d426596a39500d2..1cd966993b077b34c588f2791af27da54c099a2d 100644
--- a/src/main/java/dev/orhidea/entropy/event/NonFullBlockCollideEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/NonFullBlockCollideEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
-import dev.orhidea.entropy.event.base.EntropyBooleanEvent;
+import dev.orhidea.starforge.event.base.EntropyBooleanEvent;
 
 public class NonFullBlockCollideEvent extends EntropyBooleanEvent {
     private final Entity entity;
@@ -22,3 +22,4 @@ public class NonFullBlockCollideEvent extends EntropyBooleanEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/TntFlowEvent.java b/src/main/java/dev/orhidea/starforge/event/TntFlowEvent.java
similarity index 57%
rename from src/main/java/dev/orhidea/entropy/event/TntFlowEvent.java
rename to src/main/java/dev/orhidea/starforge/event/TntFlowEvent.java
index 94a0a29d62ad1997c903e39ada8152b880365f17..2c4c6e50ea29a8a8a26c69e53bf45c8650937a4e 100644
--- a/src/main/java/dev/orhidea/entropy/event/TntFlowEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/TntFlowEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.entity.EntropyTntPrimed;
-import dev.orhidea.entropy.event.base.EntropyTntEvent;
+import dev.orhidea.starforge.entity.EntropyTntPrimed;
+import dev.orhidea.starforge.event.base.EntropyTntEvent;
 
 public class TntFlowEvent extends EntropyTntEvent {
     public TntFlowEvent(boolean value, @NotNull EntropyTntPrimed tnt) {
@@ -10,3 +10,4 @@ public class TntFlowEvent extends EntropyTntEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/TntSpreadEvent.java b/src/main/java/dev/orhidea/starforge/event/TntSpreadEvent.java
similarity index 58%
rename from src/main/java/dev/orhidea/entropy/event/TntSpreadEvent.java
rename to src/main/java/dev/orhidea/starforge/event/TntSpreadEvent.java
index 1976950da5ab1e5b3e7a403b0696164421220600..a895a9234e6f9e144ff8d0fa29f8a529a709a9cc 100644
--- a/src/main/java/dev/orhidea/entropy/event/TntSpreadEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/TntSpreadEvent.java
@@ -1,8 +1,8 @@
-package dev.orhidea.entropy.event;
+package dev.orhidea.starforge.event;
 
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.entity.EntropyTntPrimed;
-import dev.orhidea.entropy.event.base.EntropyTntEvent;
+import dev.orhidea.starforge.entity.EntropyTntPrimed;
+import dev.orhidea.starforge.event.base.EntropyTntEvent;
 
 public class TntSpreadEvent extends EntropyTntEvent {
     public TntSpreadEvent(boolean value, @NotNull EntropyTntPrimed tnt) {
@@ -10,3 +10,4 @@ public class TntSpreadEvent extends EntropyTntEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/base/EntropyBooleanEvent.java b/src/main/java/dev/orhidea/starforge/event/base/EntropyBooleanEvent.java
similarity index 94%
rename from src/main/java/dev/orhidea/entropy/event/base/EntropyBooleanEvent.java
rename to src/main/java/dev/orhidea/starforge/event/base/EntropyBooleanEvent.java
index 42c8289f43acf021674a5d85b2875a8cebb62117..a7ca7727da5176bc4a0a7f06bed692c57b789da4 100644
--- a/src/main/java/dev/orhidea/entropy/event/base/EntropyBooleanEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/base/EntropyBooleanEvent.java
@@ -1,4 +1,4 @@
-package dev.orhidea.entropy.event.base;
+package dev.orhidea.starforge.event.base;
 
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.Event;
@@ -34,3 +34,4 @@ public abstract class EntropyBooleanEvent extends Event implements Cancellable {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/base/EntropyExplosionEvent.java b/src/main/java/dev/orhidea/starforge/event/base/EntropyExplosionEvent.java
similarity index 93%
rename from src/main/java/dev/orhidea/entropy/event/base/EntropyExplosionEvent.java
rename to src/main/java/dev/orhidea/starforge/event/base/EntropyExplosionEvent.java
index a5ca2831f3c4d6986469810b2df1b5b68c66faee..1303910bb3cbd801761ad5e0dbef53c48c76ece1 100644
--- a/src/main/java/dev/orhidea/entropy/event/base/EntropyExplosionEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/base/EntropyExplosionEvent.java
@@ -1,4 +1,4 @@
-package dev.orhidea.entropy.event.base;
+package dev.orhidea.starforge.event.base;
 
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
@@ -23,3 +23,4 @@ public abstract class EntropyExplosionEvent extends EntropyBooleanEvent {
     }
 }
 
+
diff --git a/src/main/java/dev/orhidea/entropy/event/base/EntropyFallingBlockEvent.java b/src/main/java/dev/orhidea/starforge/event/base/EntropyFallingBlockEvent.java
similarity index 81%
rename from src/main/java/dev/orhidea/entropy/event/base/EntropyFallingBlockEvent.java
rename to src/main/java/dev/orhidea/starforge/event/base/EntropyFallingBlockEvent.java
index 3631971ea44fa6c32555ef7123f1b82a6d6982cf..9f1a38624a703ce4c1aad323eba9d30c1fa9fbb3 100644
--- a/src/main/java/dev/orhidea/entropy/event/base/EntropyFallingBlockEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/base/EntropyFallingBlockEvent.java
@@ -1,7 +1,7 @@
-package dev.orhidea.entropy.event.base;
+package dev.orhidea.starforge.event.base;
 
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.entity.EntropyFallingBlock;
+import dev.orhidea.starforge.entity.EntropyFallingBlock;
 
 public abstract class EntropyFallingBlockEvent extends EntropyBooleanEvent {
     private final EntropyFallingBlock fallingBlock;
@@ -15,3 +15,4 @@ public abstract class EntropyFallingBlockEvent extends EntropyBooleanEvent {
         return this.fallingBlock;
     }
 }
+
diff --git a/src/main/java/dev/orhidea/entropy/event/base/EntropyTntEvent.java b/src/main/java/dev/orhidea/starforge/event/base/EntropyTntEvent.java
similarity index 78%
rename from src/main/java/dev/orhidea/entropy/event/base/EntropyTntEvent.java
rename to src/main/java/dev/orhidea/starforge/event/base/EntropyTntEvent.java
index 0e937fa30d26d4320c3c4c740fd4239fa6fdece3..35948b54ac72d8f5252b9a3636c7ed8e26500696 100644
--- a/src/main/java/dev/orhidea/entropy/event/base/EntropyTntEvent.java
+++ b/src/main/java/dev/orhidea/starforge/event/base/EntropyTntEvent.java
@@ -1,7 +1,7 @@
-package dev.orhidea.entropy.event.base;
+package dev.orhidea.starforge.event.base;
 
 import org.jetbrains.annotations.NotNull;
-import dev.orhidea.entropy.entity.EntropyTntPrimed;
+import dev.orhidea.starforge.entity.EntropyTntPrimed;
 
 public abstract class EntropyTntEvent extends EntropyBooleanEvent {
     private final EntropyTntPrimed tnt;
@@ -16,3 +16,4 @@ public abstract class EntropyTntEvent extends EntropyBooleanEvent {
     }
 }
 
+
