From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 16:10:06 +0100
Subject: [PATCH] Added Pufferfish Credit


diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index d4d5cabd1d0e5e7b255c2fede48bf2d0c6d1f4fd..58ad44e1004fc20fb76d3076c4eff45511007810 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -53,6 +53,7 @@ public class EntropyConfig {
                 "====================================",
                 "Entropy Cannoning Jar by FragmentMC",
                 "Support: http://fragmentmc.xyz",
+                "Pufferfish Patches: https://github.com/pufferfish-gg/Pufferfish",
                 "====================================");
 
         for (Method method : EntropyConfig.class.getDeclaredMethods()) {
