From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 17:16:14 +0100
Subject: [PATCH] Refactor Config Categories


diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 0a71716a82f0a26dcb656809e9882a988665903e..2e00587cd0a0b42d3e9f71c2397abb67c8bae2d9 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -198,9 +198,9 @@ public class EntropyConfig {
     public static Map<String, Integer> projectileTimeouts;
 	private static void projectileTimeouts() {
         // Set some defaults
-        getInt("entity_timeouts.SNOWBALL", -1);
-        getInt("entity_timeouts.LLAMA_SPIT", -1);
-        setComment("entity_timeouts",
+        getInt("pufferfish.entity_timeouts.SNOWBALL", -1);
+        getInt("pufferfish.entity_timeouts.LLAMA_SPIT", -1);
+        setComment("pufferfish.entity_timeouts",
                 "These values define a entity's maximum lifespan. If an",
                 "entity is in this list and it has survived for longer than",
                 "that number of ticks, then it will be removed. Setting a value to",
@@ -208,13 +208,13 @@ public class EntropyConfig {
 
         for (EntityType<?> entityType : Registry.ENTITY_TYPE) {
             String type = EntityType.getKey(entityType).getPath().toUpperCase(Locale.ROOT);
-            entityType.ttl = config.getInt("entity_timeouts." + type, -1);
+            entityType.ttl = config.getInt("pufferfish.entity_timeouts." + type, -1);
         }
     }
 
     public static boolean throttleInactiveGoalSelectorTick;
 	private static void inactiveGoalSelectorThrottle() {
-        		getBoolean("inactive-goal-selector-throttle", "inactive-goal-selector-disable", true,
+        		getBoolean("pufferfish.inactive-goal-selector-throttle", true,
                         "Throttles the AI goal selector in entity inactive ticks.",
                         "This can improve performance by a few percent, but has minor gameplay implications.");
     }
