From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 17:56:34 +0100
Subject: [PATCH] Move Waterflow to Entropy Config


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 5982dda61e07f1661b0a68d0ba1fcc1122e8d428..cc75a73883f79999c756300ab35aae09d8f3f1a0 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -9,22 +9,13 @@ import io.papermc.paper.configuration.legacy.MaxEntityCollisionsInitializer;
 import io.papermc.paper.configuration.legacy.RequiresSpigotInitialization;
 import io.papermc.paper.configuration.legacy.SpawnLoadedRangeInitializer;
 import io.papermc.paper.configuration.transformation.world.FeatureSeedsGeneration;
-import io.papermc.paper.configuration.type.BooleanOrDefault;
-import io.papermc.paper.configuration.type.DoubleOrDefault;
-import io.papermc.paper.configuration.type.Duration;
-import io.papermc.paper.configuration.type.EngineMode;
-import io.papermc.paper.configuration.type.IntOr;
+import io.papermc.paper.configuration.type.*;
 import io.papermc.paper.configuration.type.fallback.ArrowDespawnRate;
 import io.papermc.paper.configuration.type.fallback.AutosavePeriod;
 import it.unimi.dsi.fastutil.objects.Reference2IntMap;
 import it.unimi.dsi.fastutil.objects.Reference2IntOpenHashMap;
 import it.unimi.dsi.fastutil.objects.Reference2LongMap;
 import it.unimi.dsi.fastutil.objects.Reference2LongOpenHashMap;
-import java.util.Arrays;
-import java.util.List;
-import java.util.Map;
-import java.util.function.Function;
-import java.util.stream.Collectors;
 import net.minecraft.Util;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
@@ -44,6 +35,12 @@ import org.spongepowered.configurate.objectmapping.ConfigSerializable;
 import org.spongepowered.configurate.objectmapping.meta.Required;
 import org.spongepowered.configurate.objectmapping.meta.Setting;
 
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
 @SuppressWarnings({"FieldCanBeLocal", "FieldMayBeFinal", "NotNullFieldNotInitialized", "InnerClassMayBeStatic"})
 public class WorldConfiguration extends ConfigurationPart {
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -367,7 +364,6 @@ public class WorldConfiguration extends ConfigurationPart {
     public class Fixes extends ConfigurationPart {
         public boolean fixItemsMergingThroughWalls = false;
         public boolean disableUnloadedChunkEnderpearlExploit = true;
-        public boolean preventTntFromMovingInWater = false;
         public boolean splitOverstackedLoot = true;
         public boolean fixCuringZombieVillagerDiscountExploit = true;
         public IntOr.Disabled fallingBlockHeightNerf = IntOr.Disabled.DISABLED;
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 438de71a17d7bfaed8fbb1b4f7aa7f53538603a2..5ff0adb741c1c326d37e315c71b130105932f143 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -1,6 +1,5 @@
 package net.minecraft.world.entity.item;
 
-import javax.annotation.Nullable;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
@@ -8,15 +7,13 @@ import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.MoverType;
-import net.minecraft.world.entity.Pose;
+import net.minecraft.world.entity.*;
 import net.minecraft.world.level.Explosion;
 import net.minecraft.world.level.Level;
-import org.bukkit.event.entity.ExplosionPrimeEvent; // CraftBukkit
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+import xyz.fragmentmc.entropy.EntropyConfig;
+
+import javax.annotation.Nullable;
 
 public class PrimedTnt extends Entity {
 
@@ -98,7 +95,7 @@ public class PrimedTnt extends Entity {
         }
 
         // Paper start - Optional prevent TNT from moving in water
-        if (!this.isRemoved() && this.wasTouchingWater && this.level.paperConfig().fixes.preventTntFromMovingInWater) {
+        if (!this.isRemoved() && this.wasTouchingWater && EntropyConfig.disableWaterPushingTNT) { // Entropy - Move Waterflow to Entropy Config
             /*
              * Author: Jedediah Smith <jedediah@silencegreys.com>
              */
@@ -177,7 +174,7 @@ public class PrimedTnt extends Entity {
     // Paper start - Optional prevent TNT from moving in water
     @Override
     public boolean isPushedByFluid() {
-        return !level.paperConfig().fixes.preventTntFromMovingInWater && super.isPushedByFluid();
+        return !EntropyConfig.disableWaterPushingTNT && super.isPushedByFluid();  // Entropy - Move Waterflow to Entropy Config
     }
     // Paper end
 }
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index b5c2cfc9d429e07128cdac59961f9a64315cdeb7..e28631afd427eb08a7282964cc605808058b8c63 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -227,6 +227,14 @@ public class EntropyConfig {
         setComment("pufferfish.misc", "Settings for things that don't belong elsewhere");
     }
 
+    public static boolean disableWaterPushingTNT;
+    private static void cannoningConfiguration() {
+        disableWaterPushingTNT = getBoolean("entropy.cannons.disable-water-pushing-tnt", true,
+                "Disable TNT getting pushed by water (aka. Waterflow).");
+
+        setComment("entropy.cannons", "Settings related to cannoning");
+    }
+
     public static boolean disableChatMessageVerification;
     private static void disableChatMessageVerification() {
         disableChatMessageVerification = getBoolean("entropy.disable-chat-message-verification", true,
