From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 19:01:13 +0100
Subject: [PATCH] tnt and falling block chunk loading


diff --git a/src/main/java/io/papermc/paper/util/CollisionUtil.java b/src/main/java/io/papermc/paper/util/CollisionUtil.java
index a87f6380b2c387fb0cdd40d5087b5c93492e3c88..de780f8b49cb087e65114dbcb6cc73a49cef087a 100644
--- a/src/main/java/io/papermc/paper/util/CollisionUtil.java
+++ b/src/main/java/io/papermc/paper/util/CollisionUtil.java
@@ -2,9 +2,8 @@ package io.papermc.paper.util;
 
 import io.papermc.paper.voxel.AABBVoxelShape;
 import net.minecraft.core.BlockPos;
-import net.minecraft.server.level.ServerChunkCache;
-import net.minecraft.server.level.ServerLevel;
-import net.minecraft.server.level.WorldGenRegion;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.*;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.item.Item;
@@ -14,17 +13,14 @@ import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.PalettedContainer;
-import net.minecraft.world.level.material.FlowingFluid;
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
-import net.minecraft.world.phys.shapes.ArrayVoxelShape;
-import net.minecraft.world.phys.shapes.CollisionContext;
-import net.minecraft.world.phys.shapes.EntityCollisionContext;
-import net.minecraft.world.phys.shapes.Shapes;
-import net.minecraft.world.phys.shapes.VoxelShape;
+import net.minecraft.world.phys.shapes.*;
+
 import java.util.List;
 import java.util.function.BiPredicate;
 import java.util.function.Predicate;
@@ -482,6 +478,12 @@ public final class CollisionUtil {
                 } else {
                     chunk = loadChunks ? chunkProvider.getChunk(currChunkX, currChunkZ, true) : chunkProvider.getChunkAtIfLoadedImmediately(currChunkX, currChunkZ);
                 }
+                // Entropy start - tnt and falling block chunk loading
+                if (loadChunks && ((LevelChunk)chunk).level.getChunkSource().isChunkLoaded(currChunkX, currChunkZ)) {
+                    DistanceManager dm = ((LevelChunk) chunk).level.getChunkSource().chunkMap.distanceManager;
+                    dm.addTicket(TicketType.ENTROPY_LOAD, chunk.getPos(), 31, MinecraftServer.currentTick);
+                }
+                // Entropy end
 
                 if (chunk == null) {
                     if (collidesWithUnloaded) {
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index eab65804b78da0e4c4af54405da1dff108c53e53..f23398c0736197fcd26ef930be535716f26e75b2 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -109,6 +109,7 @@ import org.bukkit.event.weather.LightningStrikeEvent;
 import org.bukkit.event.world.GenericGameEvent;
 import org.bukkit.event.world.TimeSkipEvent;
 import org.slf4j.Logger;
+import xyz.fragmentmc.entropy.EntropyConfig;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
@@ -578,6 +579,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     public void tick(BooleanSupplier shouldKeepTicking) {
+        // Entropy start - tnt and falling block chunk loading
+        DistanceManager dm = this.getChunkSource().chunkMap.distanceManager;
+        dm.removeAllTicketsFor(TicketType.ENTROPY_LOAD, 31, MinecraftServer.currentTick - EntropyConfig.chunkLoadAliveTicks);
+        // Entropy end
+
         // Paper start - optimise checkDespawn
         this.playersAffectingSpawning.clear();
         for (ServerPlayer player : this.players) {
diff --git a/src/main/java/net/minecraft/server/level/TicketType.java b/src/main/java/net/minecraft/server/level/TicketType.java
index e9bc61590d33dc341074371859ceec54599e6c47..c1f65ecc499fe64bbe99a5bd7d95eeb88f56b413 100644
--- a/src/main/java/net/minecraft/server/level/TicketType.java
+++ b/src/main/java/net/minecraft/server/level/TicketType.java
@@ -1,14 +1,16 @@
 package net.minecraft.server.level;
 
-import java.util.Comparator;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Vec3i;
 import net.minecraft.util.Unit;
 import net.minecraft.world.level.ChunkPos;
 
+import java.util.Comparator;
+
 public class TicketType<T> {
     public static final TicketType<Long> FUTURE_AWAIT = create("future_await", Long::compareTo); // Paper
     public static final TicketType<Long> ASYNC_LOAD = create("async_load", Long::compareTo); // Paper
+    public static final TicketType<Integer> ENTROPY_LOAD = create("entropy_load", Integer::compareTo); // Entropy
 
     private final String name;
     private final Comparator<T> comparator;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c5547c4a86d2e2db7f34430617350371208d19e4..bbca3a09bdc56f7c9f525e2921ac1a1aa900d19d 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -46,7 +46,9 @@ import net.minecraft.world.Nameable;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.animal.AbstractFish;
 import net.minecraft.world.entity.animal.Animal;
+import net.minecraft.world.entity.item.FallingBlockEntity;
 import net.minecraft.world.entity.item.ItemEntity;
+import net.minecraft.world.entity.item.PrimedTnt;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.vehicle.Boat;
 import net.minecraft.world.item.Item;
@@ -1346,8 +1348,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 }
             }
 
-            io.papermc.paper.util.CollisionUtil.getCollisions(world, this, collisionBox, potentialCollisions, false, this.level.paperConfig().chunks.preventMovingIntoUnloadedChunks,
+            // Entropy start - tnt and falling block chunk loading
+            boolean load = (EntropyConfig.tntLoadChunks && this instanceof PrimedTnt) || (EntropyConfig.fallingBlockLoadChunks && this instanceof FallingBlockEntity);
+            io.papermc.paper.util.CollisionUtil.getCollisions(world, this, collisionBox, potentialCollisions, load, this.level.paperConfig().chunks.preventMovingIntoUnloadedChunks,
                 false, false, null, null);
+            // Entropy end
 
             if (io.papermc.paper.util.CollisionUtil.isCollidingWithBorderEdge(world.getWorldBorder(), collisionBox)) {
                 io.papermc.paper.util.CollisionUtil.addBoxesToIfIntersects(world.getWorldBorder().getCollisionShape(), collisionBox, potentialCollisions);
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 6e2189f5c5ec78e4292dc018a54ebe8370739032..38b75351cded6c73efcff193c487a0ef735e2488 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -233,6 +233,9 @@ public class EntropyConfig {
     public static boolean explodeWaterloggedBlocks;
     public static boolean explodeLava;
     public static boolean skipRayOrigin;
+    public static boolean tntLoadChunks;
+    public static boolean fallingBlockLoadChunks;
+    public static int chunkLoadAliveTicks;
     private static void cannoningConfiguration() {
         disableWaterPushingTNT = getBoolean("entropy.cannons.disable-water-pushing-tnt", true,
                 "Disable TNT getting pushed by water (aka. Waterflow).");
@@ -248,6 +251,14 @@ public class EntropyConfig {
                 "Use this for non-regen servers.");
         skipRayOrigin= getBoolean("entropy.cannons.skip-ray-origin", true,
                 "Skip the origin of explosion rays, as it gets checked with every ray.");
+        tntLoadChunks= getBoolean("entropy.cannons.chunks.tnt-loads-chunks", true,
+        "If TNT tries to travel through an unloaded chunk that chunk will be loaded.");
+        fallingBlockLoadChunks= getBoolean("entropy.cannons.chunks.fallingblock-loads-chunks", true,
+                "If a FallingBlock tries to travel through an unloaded chunk that chunk will be loaded.");
+        chunkLoadAliveTicks= getInt("entropy.cannons.chunks.load-alive-ticks", 2,
+                "For how many ticks a chunk that got loaded by tnt or a falling block should stay loaded.",
+                "If there is performance issues with fast shooting cannons loading chunks this can be set to ~100 ticks.");
+
 
         setComment("entropy.cannons", "Settings related to cannoning");
     }
