From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 22:28:59 +0100
Subject: [PATCH] merge cannoning entities


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index f9851537249c43d86528d540cebbb8a9d52ea150..f1f6016cbdfa4be1c7169f866ce9af2f5e08e2c9 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -48,6 +48,8 @@ import net.minecraft.world.entity.animal.WaterAnimal;
 import net.minecraft.world.entity.animal.horse.SkeletonHorse;
 import net.minecraft.world.entity.boss.EnderDragonPart;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
+import net.minecraft.world.entity.item.FallingBlockEntity;
+import net.minecraft.world.entity.item.PrimedTnt;
 import net.minecraft.world.entity.npc.Npc;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.raid.Raid;
@@ -665,6 +667,24 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 gameprofilerfiller.pop();
             }
 
+            // Entropy start - merge cannoning entities
+            if(EntropyConfig.mergeEntities) {
+                final Entity[] prev = {null};
+                entityTickList.forEach(entity -> {
+                    if (entity instanceof PrimedTnt || entity instanceof FallingBlockEntity) {
+                        if (prev[0] instanceof PrimedTnt && entity instanceof PrimedTnt && ((PrimedTnt) prev[0]).equalsMerge((PrimedTnt) entity)) {
+                            ((PrimedTnt) entity).count += ((PrimedTnt) prev[0]).count;
+                            prev[0].discard();
+                        } else if (prev[0] instanceof FallingBlockEntity && entity instanceof FallingBlockEntity && ((FallingBlockEntity) prev[0]).equalsMerge((FallingBlockEntity) entity)) {
+                            ((FallingBlockEntity) entity).count += ((FallingBlockEntity) prev[0]).count;
+                            prev[0].discard();
+                        }
+                        prev[0] = entity;
+                    }
+                });
+            }
+            // Entropy end
+
             org.spigotmc.ActivationRange.activateEntities(this); // Spigot
             timings.entityTick.startTiming(); // Spigot
             this.entityTickList.forEach((entity) -> {
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index b2d1a17867cdbaad0c6e5c2376c716f9461af124..e5358db7d0549d77222c65375c6edaef981cb6cd 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -1,9 +1,6 @@
 package net.minecraft.world.entity.item;
 
 import com.mojang.logging.LogUtils;
-import java.util.Iterator;
-import java.util.function.Predicate;
-import javax.annotation.Nullable;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -31,12 +28,7 @@ import net.minecraft.world.level.ClipContext;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.ItemLike;
 import net.minecraft.world.level.Level;
-import net.minecraft.world.level.block.AnvilBlock;
-import net.minecraft.world.level.block.Block;
-import net.minecraft.world.level.block.Blocks;
-import net.minecraft.world.level.block.ConcretePowderBlock;
-import net.minecraft.world.level.block.Fallable;
-import net.minecraft.world.level.block.FallingBlock;
+import net.minecraft.world.level.block.*;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
@@ -44,9 +36,12 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.slf4j.Logger;
 
-import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.function.Predicate;
 
 public class FallingBlockEntity extends Entity {
 
@@ -63,6 +58,8 @@ public class FallingBlockEntity extends Entity {
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Auto expire setting
 
+    public int count = 1; // Entropy - merge cannoning entities
+
     public FallingBlockEntity(EntityType<? extends FallingBlockEntity> type, Level world) {
         super(type, world);
         this.blockState = Blocks.SAND.defaultBlockState();
@@ -127,6 +124,8 @@ public class FallingBlockEntity extends Entity {
 
     @Override
     public void tick() {
+        FallingBlockEntity copy = this.clone(); // Entropy - merge cannoning entities
+
         // Paper start - fix sand duping
         if (this.isRemoved()) {
             return;
@@ -249,6 +248,15 @@ public class FallingBlockEntity extends Entity {
                             this.discard();
                             this.callOnBrokenAfterFall(block, blockposition);
                         }
+
+                        // Entropy start - merge cannoning entities
+                        if (this.count > 1) {
+                            for (int i = 0; i < this.count-1; i++) {
+                                copy.clone().tick();
+                            }
+                        }
+                        copy.discard();
+                        // Entropy end
                     }
                 }
             }
@@ -419,4 +427,27 @@ public class FallingBlockEntity extends Entity {
         this.setPos(d0, d1, d2);
         this.setStartPos(this.blockPosition());
     }
+
+    // Entropy start - merge cannoning entities
+    public boolean equalsMerge(FallingBlockEntity t) {
+        return (this.getY() == t.getY() &&
+                this.getX() == t.getX() &&
+                this.getZ() == t.getZ() &&
+                this.getDeltaMovement().x == t.getDeltaMovement().x &&
+                this.getDeltaMovement().y == t.getDeltaMovement().y &&
+                this.getDeltaMovement().z == t.getDeltaMovement().z &&
+                this.getBoundingBox().equals(t.getBoundingBox()) &&
+                this.blockState.equals(t.blockState)
+        );
+    }
+
+    @Override
+    public FallingBlockEntity clone() {
+        FallingBlockEntity neo = new FallingBlockEntity(this.level, this.getX(), this.getY(), this.getZ(), this.blockState);
+        neo.setDeltaMovement(this.getDeltaMovement());
+        neo.setBoundingBox(this.getBoundingBox());
+        neo.time = this.time;
+        return neo;
+    }
+    // Entropy end
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 2622b937e87dbfb6f6ca812ac6868c960c852d86..7ab531afbe79d51fb14ecf0f211704bf87345c2d 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -24,6 +24,8 @@ public class PrimedTnt extends Entity {
     public float yield = 4; // CraftBukkit - add field
     public boolean isIncendiary = false; // CraftBukkit - add field
 
+    public int count = 1; // Entropy - merge cannoning entities
+
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level world) {
         super(type, world);
         this.blocksBuilding = true;
@@ -63,6 +65,7 @@ public class PrimedTnt extends Entity {
 
     @Override
     public void tick() {
+        PrimedTnt copy = this.clone(); // Entropy - merge cannoning entities
         if (level.spigotConfig.maxTntTicksPerTick > 0 && ++level.spigotConfig.currentPrimedTnt > level.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
         if (!this.isNoGravity()) {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0D, -0.04D, 0.0D));
@@ -80,6 +83,15 @@ public class PrimedTnt extends Entity {
             this.setDeltaMovement(this.getDeltaMovement().multiply(0.7D, -0.5D, 0.7D));
         }
 
+        // Entropy start - merge cannoning entities
+        if (this.getFuse() - 1 <= 0  && this.count > 1) {
+            for (int x = 0; x < this.count -1; x++) {
+                PrimedTnt merged = copy.clone();
+                merged.tick();
+            }
+        }
+        // Entropy end
+
         int i = this.getFuse() - 1;
 
         this.setFuse(i);
@@ -181,4 +193,28 @@ public class PrimedTnt extends Entity {
         return !EntropyConfig.disableWaterPushingTNT && super.isPushedByFluid();  // Entropy - Move Waterflow to Entropy Config
     }
     // Paper end
+
+    // Entropy start - merge cannoning entities
+    public boolean equalsMerge(PrimedTnt t) {
+        return (this.getFuse() == t.getFuse() &&
+                this.getY() == t.getY() &&
+                this.getX() == t.getX() &&
+                this.getZ() == t.getZ() &&
+                this.getDeltaMovement().x == t.getDeltaMovement().x &&
+                this.getDeltaMovement().y == t.getDeltaMovement().y &&
+                this.getDeltaMovement().z == t.getDeltaMovement().z &&
+                this.getBoundingBox().equals(t.getBoundingBox())
+        );
+    }
+
+    @Override
+    public PrimedTnt clone(){
+        PrimedTnt neo = new PrimedTnt(this.level, this.getX(), this.getY(), this.getZ(), this.owner);
+        neo.setPos(this.position());
+        neo.setFuse(this.getFuse());
+        neo.setDeltaMovement(this.getDeltaMovement());
+        neo.setBoundingBox(this.getBoundingBox());
+        return neo;
+    }
+    //Entropy end
 }
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index a26be81033bf3d6679485cc545e01b1f6ed4e39a..fea4a3a715e7fe39c863abba4ae06ba34ce55f42 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -255,6 +255,7 @@ public class EntropyConfig {
     public static String durabilityCheckItem;
     public static String durabilityCheckMessage;
     public static Map<String, Integer> unbreakableMap;
+    public static boolean mergeEntities;
     private static void cannoningConfiguration() {
         disableWaterPushingTNT = getBoolean("entropy.cannons.disable-water-pushing-tnt", true,
                 "Disable TNT getting pushed by water (aka. Waterflow).");
@@ -291,6 +292,8 @@ public class EntropyConfig {
                 "Message that shows when checking the durability of an unbreakable block.");
         unbreakableMap= getStringIntMap("entropy.cannons.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults(),
         "block: durability");
+        mergeEntities= getBoolean("entropy.cannons.merge-entities", true,
+                "Merge tnt and falling blocks and move them together.");
 
 
 
