From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 22:39:33 +0100
Subject: [PATCH] combined explosions


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 7ab531afbe79d51fb14ecf0f211704bf87345c2d..61d6ecf593598f17a4214e193cd446723ec52681 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -24,6 +24,7 @@ public class PrimedTnt extends Entity {
     public float yield = 4; // CraftBukkit - add field
     public boolean isIncendiary = false; // CraftBukkit - add field
 
+    public boolean canCombine = false; // Entropy - combined explosions
     public int count = 1; // Entropy - merge cannoning entities
 
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level world) {
@@ -65,7 +66,25 @@ public class PrimedTnt extends Entity {
 
     @Override
     public void tick() {
-        PrimedTnt copy = this.clone(); // Entropy - merge cannoning entities
+        // Entropy start - merge cannoning entities
+        if (this.getFuse() - 1 <= 0 && this.count > 1) {
+            PrimedTnt vtnt = this.clone();
+            vtnt.setFuse(100);
+            vtnt.tick();
+
+            if (EntropyConfig.mergeEntities &&
+                    vtnt.level.getBlockState(vtnt.blockPosition()).getMaterial().isLiquid() &&
+                    ((this.getDeltaMovement().x == 0 && this.getDeltaMovement().y == 0 && this.getDeltaMovement().z == 0) || this.distanceTo(vtnt) > 8)) {
+                this.canCombine = true;
+            } else {
+                for (int x = 0; x < this.count -1; x++) {
+                    PrimedTnt merged = this.clone();
+                    merged.tick();
+                }
+            }
+            vtnt.discard();
+        }
+        // Entropy end
         if (level.spigotConfig.maxTntTicksPerTick > 0 && ++level.spigotConfig.currentPrimedTnt > level.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
         if (!this.isNoGravity()) {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0D, -0.04D, 0.0D));
@@ -83,15 +102,6 @@ public class PrimedTnt extends Entity {
             this.setDeltaMovement(this.getDeltaMovement().multiply(0.7D, -0.5D, 0.7D));
         }
 
-        // Entropy start - merge cannoning entities
-        if (this.getFuse() - 1 <= 0  && this.count > 1) {
-            for (int x = 0; x < this.count -1; x++) {
-                PrimedTnt merged = copy.clone();
-                merged.tick();
-            }
-        }
-        // Entropy end
-
         int i = this.getFuse() - 1;
 
         this.setFuse(i);
diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 1ba996070bdc86bb3b5d46225bef03e63cf17ee7..0b938288cf84fdb3308b1e9cbb758ae8ec163eb1 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -142,6 +142,12 @@ public class Explosion {
     }
 
     public void explode() {
+        // Entropy start - combined explosions
+        int merged = 1;
+        if (this.source instanceof PrimedTnt && ((PrimedTnt) this.source).canCombine) {
+            merged = ((PrimedTnt) this.source).count;
+        }
+        // Entropy end
         // CraftBukkit start
         if (this.radius < 0.1F) {
             return;
@@ -338,7 +344,15 @@ public class Explosion {
                             d14 = entity instanceof Player && level.paperConfig().environment.disableExplosionKnockback ? 0 : ProtectionEnchantment.getExplosionKnockbackAfterDampener((LivingEntity) entity, d13); // Paper - Disable explosion knockback
                         }
 
-                        entity.setDeltaMovement(entity.getDeltaMovement().add(d8 * d14, d9 * d14, d10 * d14));
+                        // Entropy start - combined explosions
+                        double tax = 0, tay = 0, taz = 0;
+                        for (int k = 0; k < merged; k++) {
+                            tax += d8 * d14;
+                            tay += d9 * d14;
+                            taz += d10 * d14;
+                        }
+                        entity.setDeltaMovement(entity.getDeltaMovement().add(tax, tay, taz));
+                        // Entropy end
                         if (entity instanceof Player) {
                             Player entityhuman = (Player) entity;
 
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index fea4a3a715e7fe39c863abba4ae06ba34ce55f42..488ea5221b52f4a858b4100dc450cfd994955e26 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -256,6 +256,7 @@ public class EntropyConfig {
     public static String durabilityCheckMessage;
     public static Map<String, Integer> unbreakableMap;
     public static boolean mergeEntities;
+    public static boolean combinedExplosions;
     private static void cannoningConfiguration() {
         disableWaterPushingTNT = getBoolean("entropy.cannons.disable-water-pushing-tnt", true,
                 "Disable TNT getting pushed by water (aka. Waterflow).");
@@ -294,6 +295,8 @@ public class EntropyConfig {
         "block: durability");
         mergeEntities= getBoolean("entropy.cannons.merge-entities", true,
                 "Merge tnt and falling blocks and move them together.");
+        combinedExplosions= getBoolean("entropy.cannons.combined-explosions", true,
+                "Combine merged tnt explosions into one big explosion.");
 
 
 
