From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Thu, 6 Jul 2023 16:53:59 +0200
Subject: [PATCH] Entropy Command System and Config


diff --git a/src/main/java/gg/pufferfish/pufferfish/PufferfishCommand.java b/src/main/java/gg/pufferfish/pufferfish/PufferfishCommand.java
index 020368da69b9a492155f6de6297f74732f4ab6ea..50e2c21f1ca9164782d181c9f7b08dc51efd89d4 100644
--- a/src/main/java/gg/pufferfish/pufferfish/PufferfishCommand.java
+++ b/src/main/java/gg/pufferfish/pufferfish/PufferfishCommand.java
@@ -19,7 +19,7 @@ public class PufferfishCommand extends Command {
     public PufferfishCommand() {
         super("pufferfish");
         this.description = "Pufferfish related commands";
-        this.usageMessage = "/pufferfish [reload | version]";
+        this.usageMessage = "/pufferfish [reload]";
         this.setPermission("bukkit.command.pufferfish");
     }
     
@@ -30,7 +30,7 @@ public class PufferfishCommand extends Command {
     @Override
     public List<String> tabComplete(CommandSender sender, String alias, String[] args, Location location) throws IllegalArgumentException {
         if (args.length == 1) {
-            return Stream.of("reload", "version")
+            return Stream.of("reload")
               .filter(arg -> arg.startsWith(args[0].toLowerCase()))
               .collect(Collectors.toList());
         }
@@ -59,8 +59,6 @@ public class PufferfishCommand extends Command {
             console.server.reloadCount++;
 
             Command.broadcastCommandMessage(sender, prefix + "Pufferfish configuration has been reloaded.");
-        } else if (args[0].equalsIgnoreCase("version")) {
-            Command.broadcastCommandMessage(sender, prefix + "This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")");
         }
 
         return true;
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 28af96ce4f9f1a83316b6fd9e1625c71b5874c3f..75463fd35ec20546a1c8177b6280c8cd5155791f 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -224,6 +224,9 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         gg.pufferfish.pufferfish.PufferfishConfig.load(); // Pufferfish
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish
 
+        xyz.fragmentmc.entropy.EntropyConfig.load(); // Entropy
+        xyz.fragmentmc.entropy.command.EntropyCommandLoader.loadCommands(); // Entropy
+
         this.setPvpAllowed(dedicatedserverproperties.pvp);
         this.setFlightAllowed(dedicatedserverproperties.allowFlight);
         this.setMotd(dedicatedserverproperties.motd);
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..163c2ffad6cd20170c0dfe3b59f91d64898e80ec
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -0,0 +1,131 @@
+package xyz.fragmentmc.entropy;
+
+import gg.pufferfish.pufferfish.PufferfishConfig;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.configuration.MemoryConfiguration;
+import org.jetbrains.annotations.Nullable;
+import org.simpleyaml.configuration.comments.CommentType;
+import org.simpleyaml.configuration.file.YamlFile;
+import org.simpleyaml.exceptions.InvalidConfigurationException;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+
+public class EntropyConfig {
+    private static final YamlFile config = new YamlFile();
+    private static int updates = 0;
+
+    private static ConfigurationSection convertToBukkit(org.simpleyaml.configuration.ConfigurationSection section) {
+        ConfigurationSection newSection = new MemoryConfiguration();
+        for (String key : section.getKeys(false)) {
+            if (section.isConfigurationSection(key)) {
+                newSection.set(key, convertToBukkit(section.getConfigurationSection(key)));
+            } else {
+                newSection.set(key, section.get(key));
+            }
+        }
+        return newSection;
+    }
+
+    public static ConfigurationSection getConfigCopy() {
+        return convertToBukkit(config);
+    }
+
+    public static int getUpdates() {
+        return updates;
+    }
+    public static void load() throws IOException {
+        File configFile = new File("entropy.yml");
+        if (configFile.exists()) {
+            try {
+                config.load(configFile);
+            } catch (InvalidConfigurationException e) {
+                throw new IOException(e);
+            }
+        }
+        getString("info.version", "1.0");
+        setComment("info",
+                "====================================",
+                "Entropy Cannoning Jar by FragmentMC",
+                "Support: http://fragmentmc.xyz",
+                "====================================");
+
+		for (Method method : PufferfishConfig .class.getDeclaredMethods()) {
+        if (Modifier.isStatic(method.getModifiers()) && Modifier.isPrivate(method.getModifiers()) && method.getParameterCount() == 0 &&
+                method.getReturnType() == Void.TYPE && !method.getName().startsWith("lambda")) {
+            method.setAccessible(true);
+            try {
+                method.invoke(null);
+            } catch (Throwable t) {
+                MinecraftServer.LOGGER.warn("Failed to load configuration option from " + method.getName(), t);
+            }
+        }
+    }
+
+    updates++;
+
+		config.save(configFile);
+    }
+
+    private static void setComment(String key, String... comment) {
+        if (config.contains(key)) {
+            config.setComment(key, String.join("\n", comment), CommentType.BLOCK);
+        }
+    }
+
+    private static void ensureDefault(String key, Object defaultValue, String... comment) {
+        if (!config.contains(key)) {
+            config.set(key, defaultValue);
+            config.setComment(key, String.join("\n", comment), CommentType.BLOCK);
+        }
+    }
+
+    private static boolean getBoolean(String key, boolean defaultValue, String... comment) {
+        return getBoolean(key, null, defaultValue, comment);
+    }
+
+    private static boolean getBoolean(String key, @Nullable String oldKey, boolean defaultValue, String... comment) {
+        ensureDefault(key, defaultValue, comment);
+        return config.getBoolean(key, defaultValue);
+    }
+
+    private static int getInt(String key, int defaultValue, String... comment) {
+        return getInt(key, null, defaultValue, comment);
+    }
+
+    private static int getInt(String key, @Nullable String oldKey, int defaultValue, String... comment) {
+        ensureDefault(key, defaultValue, comment);
+        return config.getInt(key, defaultValue);
+    }
+
+    private static double getDouble(String key, double defaultValue, String... comment) {
+        return getDouble(key, null, defaultValue, comment);
+    }
+
+    private static double getDouble(String key, @Nullable String oldKey, double defaultValue, String... comment) {
+        ensureDefault(key, defaultValue, comment);
+        return config.getDouble(key, defaultValue);
+    }
+
+    private static String getString(String key, String defaultValue, String... comment) {
+        return getOldString(key, null, defaultValue, comment);
+    }
+
+    private static String getOldString(String key, @Nullable String oldKey, String defaultValue, String... comment) {
+        ensureDefault(key, defaultValue, comment);
+        return config.getString(key, defaultValue);
+    }
+
+    private static List<String> getStringList(String key, List<String> defaultValue, String... comment) {
+        return getStringList(key, null, defaultValue, comment);
+    }
+
+    private static List<String> getStringList(String key, @Nullable String oldKey, List<String> defaultValue, String... comment) {
+        ensureDefault(key, defaultValue, comment);
+        return config.getStringList(key);
+    }
+}
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java b/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java
index a33800cc4612ef58061169394038a4c3abfcb354..55faa5a4a1bd126bdfb081a7d098636708adaea2 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java
@@ -15,6 +15,6 @@ public class EntropyVersionFetcher implements VersionFetcher {
 
     @Override
     public @NotNull Component getVersionMessage(@NotNull String serverVersion) {
-        return Component.text("> Entropy Cannoning Jar by FragmentMC ("+ serverVersion + ") <\nGet this jar with the \"Insider\" Tier at \"https://www.patreon.com/jesusmc\"!");
+        return Component.text("> Entropy Cannoning Jar by FragmentMC <\nGet this jar with the \"Insider\" Tier at \"https://www.patreon.com/jesusmc\"!");
     }
 }
diff --git a/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandBase.java b/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandBase.java
new file mode 100644
index 0000000000000000000000000000000000000000..b39639008a1ca80bfb6d571444601e9f8247809c
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandBase.java
@@ -0,0 +1,47 @@
+package xyz.fragmentmc.entropy.command;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.JoinConfiguration;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+import org.w3c.dom.Text;
+
+import java.util.Arrays;
+
+public abstract class EntropyCommandBase extends Command {
+    private final EntropyCommandData commandData;
+
+    protected final TextComponent prefix = (TextComponent) Component.join(
+            JoinConfiguration.noSeparators(),
+            Component.text("Entropy >", NamedTextColor.GOLD, TextDecoration.BOLD),
+            Component.text(" ", NamedTextColor.WHITE));
+    public EntropyCommandData getCommandData() {
+        return commandData;
+    }
+
+    @Override
+    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
+        if (!testPermission(sender)) return true;
+        if (args.length != 1) {
+            sender.sendMessage(prefix.toBuilder().append(Component.text("Usage: " + usageMessage)).build());
+            return true;
+            }
+        return run(sender, commandLabel, args);
+    }
+
+    protected EntropyCommandBase() {
+        super("");
+        commandData = this.getClass().getAnnotation(EntropyCommandData.class);
+        this.setName(commandData.name());
+        this.setDescription(commandData.description());
+        this.setUsage(commandData.usageMessage());
+        this.setAliases(Arrays.stream(commandData.aliases()).toList());
+        this.setPermission(commandData.permission());
+    }
+
+    protected abstract boolean run(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args);
+}
diff --git a/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandData.java b/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandData.java
new file mode 100644
index 0000000000000000000000000000000000000000..7f1936d2e2365db28a52558531b8ce02bcfb444e
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandData.java
@@ -0,0 +1,13 @@
+package xyz.fragmentmc.entropy.command;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface EntropyCommandData {
+    String name() default "default";
+    String description() default "This is the default command description";
+    String usageMessage() default "/default";
+    String[] aliases() default {};
+    String permission() default "entropy.command.default";
+}
diff --git a/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandLoader.java b/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandLoader.java
new file mode 100644
index 0000000000000000000000000000000000000000..559ce60ddedddf81ad9700fa8996ede29b353cfd
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/command/EntropyCommandLoader.java
@@ -0,0 +1,15 @@
+package xyz.fragmentmc.entropy.command;
+
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.command.Command;
+import xyz.fragmentmc.entropy.command.commands.EntropyCommand;
+
+public class EntropyCommandLoader {
+    public static void loadCommands() {
+        load(new EntropyCommand());
+    }
+
+    private static void load(Command command) {
+        MinecraftServer.getServer().server.getCommandMap().register(command.getName(), command.getName(), command);
+    }
+}
diff --git a/src/main/java/xyz/fragmentmc/entropy/command/commands/EntropyCommand.java b/src/main/java/xyz/fragmentmc/entropy/command/commands/EntropyCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..9da56a9f01361e616b51063613bf41b49607c324
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/command/commands/EntropyCommand.java
@@ -0,0 +1,52 @@
+package xyz.fragmentmc.entropy.command.commands;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+import xyz.fragmentmc.entropy.EntropyConfig;
+import xyz.fragmentmc.entropy.command.EntropyCommandBase;
+import xyz.fragmentmc.entropy.command.EntropyCommandData;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@EntropyCommandData(name = "entropy", description = "Entropy related commands", usageMessage = "/entropy [reload | version]", permission = "entropy.command.entropy")
+public class EntropyCommand extends EntropyCommandBase {
+    @Override
+    public @NotNull List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, String[] args, Location location) throws IllegalArgumentException {
+        if (args.length == 1) {
+            return Stream.of("reload", "version").filter(arg -> arg.startsWith(args[0].toLowerCase())).collect(Collectors.toList());
+        }
+        return Collections.emptyList();
+    }
+
+
+    @Override
+    protected boolean run(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
+        if (args[0].equalsIgnoreCase("reload")) {
+            MinecraftServer console = MinecraftServer.getServer();
+            try {
+                EntropyConfig.load();
+            } catch (IOException e) {
+                sender.sendMessage(Component.text("Failed to reload.", NamedTextColor.RED));
+                e.printStackTrace();
+                return true;
+            }
+            console.server.reloadCount++;
+
+            Command.broadcastCommandMessage(sender, prefix.toBuilder().append(Component.text("Entropy configuration has been reloaded.")).build());
+        } else if (args[0].equalsIgnoreCase("version")) {
+            Command.broadcastCommandMessage(sender, prefix.toBuilder().append(Component.text("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")")).build());
+        }
+
+        return true;
+    }
+}
