From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Fri, 11 Aug 2023 16:53:48 +0200
Subject: [PATCH] fix powdered snow merging


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index dd1b358e2a07e327d642286ae5f81f3c6e9a9f56..8876b1eaebb69534222d3fc1bca1fba5b257f084 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -206,4 +206,31 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         return !EventUtil.sendCancellable(level, new TntFlowEvent(EntropyConfig.disableTNTFlow, new CraftEntropyTntPrimed(level.getCraftServer(), this))).isCancelled() && super.isPushedByFluid(); // Entropy
     }
     // Paper end - Option to prevent TNT from moving in water
+    
+    public boolean equalsMerge(PrimedTnt t) {
+        return (this.getFuse() == t.getFuse() &&
+                this.getY() == t.getY() &&
+                this.getX() == t.getX() &&
+                this.getZ() == t.getZ() &&
+                this.getDeltaMovement().x == t.getDeltaMovement().x &&
+                this.getDeltaMovement().y == t.getDeltaMovement().y &&
+                this.getDeltaMovement().z == t.getDeltaMovement().z &&
+                this.getBoundingBox().equals(t.getBoundingBox()) &&
+                (this.fallDistance > 2.5f || this.fallDistance == 0 || this.fallDistance == t.fallDistance)
+        );
+    }
+
+    @Override
+    public PrimedTnt clone() {
+        PrimedTnt neo = new PrimedTnt(this.level, this.getX(), this.getY(), this.getZ(), this.owner);
+        neo.setDeltaMovement(this.getDeltaMovement());
+        neo.setBoundingBox(this.getBoundingBox());
+        neo.setFuse(this.getFuse());
+        if(this.getOriginVector() != null && this.getOriginWorld() != null) {
+            neo.setOriginWorld(this.getOriginWorld());
+            neo.setOriginVector(this.getOriginVector());
+        }
+        return neo;
+    }
+    // Entropy end
 }
