From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 6 Dec 2022 22:46:48 +0100
Subject: [PATCH] reformatted config


diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 488ea5221b52f4a858b4100dc450cfd994955e26..deefedb71c635699c186c91d9792bbbeb318f9c2 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -241,9 +241,11 @@ public class EntropyConfig {
 
     public static boolean disableWaterPushingTNT;
     public static boolean disableTNTSpread;
-    public static boolean explosionWaterCheck;
     public static boolean explodeWaterloggedBlocks;
     public static boolean explodeLava;
+    public static boolean mergeEntities;
+    public static boolean combinedExplosions;
+    public static boolean explosionWaterCheck;
     public static boolean skipRayOrigin;
     public static boolean tntLoadChunks;
     public static boolean fallingBlockLoadChunks;
@@ -255,54 +257,47 @@ public class EntropyConfig {
     public static String durabilityCheckItem;
     public static String durabilityCheckMessage;
     public static Map<String, Integer> unbreakableMap;
-    public static boolean mergeEntities;
-    public static boolean combinedExplosions;
     private static void cannoningConfiguration() {
-        disableWaterPushingTNT = getBoolean("entropy.cannons.disable-water-pushing-tnt", true,
+        disableWaterPushingTNT = getBoolean("entropy.patches.disable-water-pushing-tnt", true,
                 "Disable TNT getting pushed by water (aka. Waterflow).");
-        disableTNTSpread = getBoolean("entropy.cannons.disable-tnt-spread", true,
+        disableTNTSpread = getBoolean("entropy.patches.disable-tnt-spread", true,
                 "Disable the random velocity TNT gets when ignited.");
-        explosionWaterCheck= getBoolean("entropy.cannons.explosion-submerged-check", true,
-                "Disable block breaking checks when tnt explodes while submerged.");
-        explodeWaterloggedBlocks= getBoolean("entropy.cannons.explode-waterlogged-blocks", true,
+        explodeWaterloggedBlocks= getBoolean("entropy.patches.explode-waterlogged-blocks", true,
                 "Waterlogged blocks will be destroyed by explosions.",
                 "This can be used to prevent unraidable bases.");
-        explodeLava= getBoolean("entropy.cannons.explode-lava", false,
+        explodeLava= getBoolean("entropy.patches.explode-lava", false,
                 "Lava will be destroyed by explosions.",
                 "Use this for non-regen servers.");
-        skipRayOrigin= getBoolean("entropy.cannons.skip-ray-origin", true,
+
+        mergeEntities= getBoolean("entropy.optimizations.merge-entities", true,
+                "Merge tnt and falling blocks and move them together.");
+        combinedExplosions= getBoolean("entropy.optimizations.combined-explosions", true,
+                "Combine merged tnt explosions into one big explosion.");
+        explosionWaterCheck= getBoolean("entropy.optimizations.explosion-submerged-check", true,
+                "Disable block breaking checks when tnt explodes while submerged.");
+        skipRayOrigin= getBoolean("entropy.optimizations.skip-ray-origin", true,
                 "Skip the origin of explosion rays, as it gets checked with every ray.");
-        tntLoadChunks= getBoolean("entropy.cannons.chunks.tnt-loads-chunks", true,
+        tntLoadChunks= getBoolean("entropy.chunks.tnt-loads-chunks", true,
         "If TNT tries to travel through an unloaded chunk that chunk will be loaded.");
-        fallingBlockLoadChunks= getBoolean("entropy.cannons.chunks.fallingblock-loads-chunks", true,
+        fallingBlockLoadChunks= getBoolean("entropy.chunks.fallingblock-loads-chunks", true,
                 "If a FallingBlock tries to travel through an unloaded chunk that chunk will be loaded.");
-        chunkLoadAliveTicks= getInt("entropy.cannons.chunks.load-alive-ticks", 2,
+        chunkLoadAliveTicks= getInt("entropy.chunks.load-alive-ticks", 2,
                 "For how many ticks a chunk that got loaded by tnt or a falling block should stay loaded.",
                 "If there is performance issues with fast shooting cannons loading chunks this can be set to ~100 ticks.");
-        explosionProofItems= getBoolean("entropy.cannons.explosion-proof-items.enabled", true,
+        explosionProofItems= getBoolean("entropy..explosion-proof-items.enabled", true,
                 "Enable a list of items that can not be blown up by explosions.");
-        explosionProofList= getStringList("entropy.cannons.explosion-proof-items.items", List.of("spawner"));
+        explosionProofList= getStringList("entropy.explosion-proof-items.items", List.of("spawner"));
         ExplosionProofItems.loadFromConfig();
-        unbreakableBreaker= getBoolean("entropy.cannons.unbreakable-breaker.enabled", true,
+        unbreakableBreaker= getBoolean("entropy.unbreakable-breaker.enabled", true,
                 "Assigns unbreakable blocks like obsidian a durability so they can be broken by tnt.");
-        unbreakableCreeper= getBoolean("entropy.cannons.unbreakable-breaker.allow-creepers", false,
+        unbreakableCreeper= getBoolean("entropy.unbreakable-breaker.allow-creepers", false,
                 "Should unbreakable blocks be able to be broken by creepers.");
-        durabilityCheckItem= getString("entropy.cannons.unbreakable-breaker.durability-check-item", "potato",
+        durabilityCheckItem= getString("entropy.unbreakable-breaker.durability-check-item", "potato",
                 "Item used to check an unbreakable blocks durability.");
-        durabilityCheckMessage= getString("entropy.cannons.unbreakable-breaker.durability-check-message", "&a[+] &7This block has %dura%/%max% durability.",
+        durabilityCheckMessage= getString("entropy.unbreakable-breaker.durability-check-message", "&a[+] &7This block has %dura%/%max% durability.",
                 "Message that shows when checking the durability of an unbreakable block.");
-        unbreakableMap= getStringIntMap("entropy.cannons.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults(),
-        "block: durability");
-        mergeEntities= getBoolean("entropy.cannons.merge-entities", true,
-                "Merge tnt and falling blocks and move them together.");
-        combinedExplosions= getBoolean("entropy.cannons.combined-explosions", true,
-                "Combine merged tnt explosions into one big explosion.");
-
-
-
-
-
-        setComment("entropy.cannons", "Settings related to cannoning");
+        unbreakableMap= getStringIntMap("entropy.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults());
+        setComment("entropy.misc.cannons", "Settings related to cannoning");
     }
 
     public static boolean disableChatMessageVerification;
