From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Fri, 7 Jul 2023 04:07:44 +0200
Subject: [PATCH] Entropy Raise Spigot TNT Per Tick


diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 5b5109e942b18418b3a3a0e2109fe4ef15045fe5..ff513e129fcb9653702f427657ccba2695f728ff 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -438,9 +438,9 @@ public class SpigotWorldConfig
     private void maxTntPerTick() {
         if ( SpigotConfig.version < 7 )
         {
-            this.set( "max-tnt-per-tick", 100 );
+            this.set( "max-tnt-per-tick", 10000 ); // Entropy
         }
-        this.maxTntTicksPerTick = this.getInt( "max-tnt-per-tick", 100 );
+        this.maxTntTicksPerTick = this.getInt( "max-tnt-per-tick", 10000 ); // Entropy
         this.log( "Max TNT Explosions: " + this.maxTntTicksPerTick );
     }
 
