From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Mon, 11 Dec 2023 15:07:15 +0100
Subject: [PATCH] branding changes


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 18bfe2fb7efad66f5fae07a30593d640c597bf77..0739662beb8d32a42a31e3cf76753f967e3bf8de 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -266,7 +266,7 @@ import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
-    private final String serverName = "Pufferfish"; // Paper // Pufferfish
+    private final String serverName = "Entropy"; // Paper // Pufferfish // Entropy
     private final String serverVersion;
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java b/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java
index 55faa5a4a1bd126bdfb081a7d098636708adaea2..bcc4831e76a05a8354ba29b2e5dca61785849bb6 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyVersionFetcher.java
@@ -15,6 +15,6 @@ public class EntropyVersionFetcher implements VersionFetcher {
 
     @Override
     public @NotNull Component getVersionMessage(@NotNull String serverVersion) {
-        return Component.text("> Entropy Cannoning Jar by FragmentMC <\nGet this jar with the \"Insider\" Tier at \"https://www.patreon.com/jesusmc\"!");
+        return Component.text("> Entropy Cannoning Jar by FragmentMC <\nGet this jar at \"http://discord.fragmentmc.xyz\"!");
     }
 }
