From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Wed, 7 Dec 2022 12:20:41 +0100
Subject: [PATCH] raise default max tnt per tick


diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 5503ad6a93d331771a0e92c0da6adedf2ac81aff..ef61136cbd2eb4acaecaceb6dbe6a001811b69f8 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -1,9 +1,10 @@
 package org.spigotmc;
 
-import java.util.List;
 import org.bukkit.Bukkit;
 import org.bukkit.configuration.file.YamlConfiguration;
 
+import java.util.List;
+
 public class SpigotWorldConfig
 {
 
@@ -430,9 +431,9 @@ public class SpigotWorldConfig
     private void maxTntPerTick() {
         if ( SpigotConfig.version < 7 )
         {
-            this.set( "max-tnt-per-tick", 100 );
+            this.set( "max-tnt-per-tick", 10000 );
         }
-        this.maxTntTicksPerTick = this.getInt( "max-tnt-per-tick", 100 );
+        this.maxTntTicksPerTick = this.getInt( "max-tnt-per-tick", 10000 );
         this.log( "Max TNT Explosions: " + this.maxTntTicksPerTick );
     }
 
