From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Fri, 9 Dec 2022 12:27:33 +0100
Subject: [PATCH] added config comments


diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 05245e97448e14d34a41b6491cedaeb6121aef40..c445b85ae406113a0ee3f1173dad7d288545efb6 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -18,7 +18,7 @@ import java.util.*;
 
 public class EntropyConfig {
     private static final YamlFile config = new YamlFile();
-    private static int updates = 0;
+    private static int updates = 1;
 
     private static ConfigurationSection convertToBukkit(org.simpleyaml.configuration.ConfigurationSection section) {
         ConfigurationSection newSection = new MemoryConfiguration();
@@ -50,7 +50,7 @@ public class EntropyConfig {
             }
         }
 
-        getString("info.version", "1.0");
+        getString("info.version", "1.1");
         setComment("info",
                 "====================================",
                 "Entropy Cannoning Jar by FragmentMC",
@@ -272,6 +272,8 @@ public class EntropyConfig {
         explodeLava= getBoolean("entropy.patches.explode-lava", false,
                 "Lava will be destroyed by explosions.",
                 "Use this for non-regen servers.");
+        setComment("entropy.patches", "Changes to game mechanics related to cannoning");
+
         mergeEntities= getBoolean("entropy.optimizations.merge-entities", true,
                 "Merge tnt and falling blocks and move them together.");
         combinedExplosions= getBoolean("entropy.optimizations.combined-explosions", true,
@@ -280,6 +282,8 @@ public class EntropyConfig {
                 "Disable block breaking checks when tnt explodes while submerged.");
         skipRayOrigin= getBoolean("entropy.optimizations.skip-ray-origin", true,
                 "Skip the origin of explosion rays, as it gets checked with every ray.");
+        setComment("entropy.optimizations", "Changes that improve cannoning performance");
+
         tntLoadChunks= getBoolean("entropy.chunks.tnt-loads-chunks", true,
         "If TNT tries to travel through an unloaded chunk that chunk will be loaded.");
         fallingBlockLoadChunks= getBoolean("entropy.chunks.fallingblock-loads-chunks", true,
@@ -287,10 +291,13 @@ public class EntropyConfig {
         chunkLoadAliveTicks= getInt("entropy.chunks.load-alive-ticks", 2,
                 "For how many ticks a chunk that got loaded by tnt or a falling block should stay loaded.",
                 "If there is performance issues with fast shooting cannons loading chunks this can be set to ~100 ticks.");
-        explosionProofItems= getBoolean("entropy..explosion-proof-items.enabled", true,
+        setComment("entropy.optimizations", "Chunk loading related settings");
+
+        explosionProofItems= getBoolean("entropy.explosion-proof-items.enabled", true,
                 "Enable a list of items that can not be blown up by explosions.");
         explosionProofList= getStringList("entropy.explosion-proof-items.items", List.of("spawner"));
         ExplosionProofItems.loadFromConfig();
+
         unbreakableBreaker= getBoolean("entropy.unbreakable-breaker.enabled", true,
                 "Assigns unbreakable blocks like obsidian a durability so they can be broken by tnt.");
         unbreakableCreeper= getBoolean("entropy.unbreakable-breaker.allow-creepers", false,
@@ -300,7 +307,6 @@ public class EntropyConfig {
         durabilityCheckMessage= getString("entropy.unbreakable-breaker.durability-check-message", "&a[+] &7This block has %dura%/%max% durability.",
                 "Message that shows when checking the durability of an unbreakable block.");
         unbreakableMap= getStringIntMap("entropy.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults());
-        setComment("entropy.misc.cannons", "Settings related to cannoning");
     }
 
     public static boolean disableChatMessageVerification;
