From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Fri, 7 Jul 2023 04:22:25 +0200
Subject: [PATCH] Entropy Range Limits


diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 5d85b201b6601ae4a65a51e76ea37d6b8a03cddb..4b5ad38524296cb5cf90ee4844377e3207b3d8c1 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -47,6 +47,7 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.util.Vector;
 import org.slf4j.Logger;
 
 // CraftBukkit start;
@@ -156,6 +157,17 @@ public class FallingBlockEntity extends Entity {
 
             this.move(MoverType.SELF, this.getDeltaMovement());
 
+            // Entropy start
+            if (EntropyConfig.cannonRangeLimit) {
+                Vector origin = this.getOriginVector();
+                if (origin != null && (
+                        Math.abs(this.position().x - origin.getX()) > EntropyConfig.cannonRangeLimitChunks * 16 ||
+                        Math.abs(this.position().z - origin.getZ()) > EntropyConfig.cannonRangeLimitChunks * 16
+                        )
+                ) this.discard();
+            }
+            // Entropy end
+
             // Paper start - fix sand duping
             if (this.isRemoved()) {
                 return;
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index c81e1d2c5e52ff148c9714b75652daebd9e89ca8..bad76f81feb5c4f8ce7cbe87403f228d493edc7f 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -17,6 +17,7 @@ import net.minecraft.world.level.Level;
 // CraftBukkit start;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
+import org.bukkit.util.Vector;
 import xyz.fragmentmc.entropy.EntropyConfig;
 import xyz.fragmentmc.entropy.entity.CraftEntropyTntPrimed;
 import xyz.fragmentmc.entropy.event.TntFlowEvent;
@@ -103,6 +104,18 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         }
 
         this.move(MoverType.SELF, this.getDeltaMovement());
+
+        // Entropy start
+        if (EntropyConfig.cannonRangeLimit) {
+            Vector origin = this.getOriginVector();
+            if (origin != null && (
+                    Math.abs(this.position().x - origin.getX()) > EntropyConfig.cannonRangeLimitChunks * 16 ||
+                            Math.abs(this.position().z - origin.getZ()) > EntropyConfig.cannonRangeLimitChunks * 16
+            )
+            ) this.discard();
+        }
+        // Entropy end
+
         // Paper start - Configurable TNT entity height nerf
         if (this.level().paperConfig().fixes.tntEntityHeightNerf.test(v -> this.getY() > v)) {
             this.discard();
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index e8114d89a3129e56c0329410a49ded63cc77cb4c..2181ddc4c385f4c92fe74b5c729342d61cf33c45 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -20,6 +20,8 @@ import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.event.player.PlayerTeleportEvent;
+import org.bukkit.util.Vector;
+import xyz.fragmentmc.entropy.EntropyConfig;
 // CraftBukkit end
 
 public class ThrownEnderpearl extends ThrowableItemProjectile {
@@ -109,6 +111,15 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
             super.tick();
         }
 
+        // Entropy start
+        if(EntropyConfig.pearlRangeLimit) {
+            Vector origin = this.getOriginVector();
+            if(origin != null && Math.sqrt(Math.pow(Math.abs(position().x - origin.getX()), 2)) + Math.sqrt(Math.pow(Math.abs(position().z - origin.getZ()), 2)) > EntropyConfig.pearlRangeLimitChunks * 16) {
+                this.setDeltaMovement(0, this.getY(), 0);
+                this.moveTo(this.xOld, this.position().y, this.yOld);
+            }
+        }
+        // Entropy end
     }
 
     @Nullable
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 22380132dd3ce1e6955b49b931b30fbbccc964f4..5d7a4fcffa2bab143b38aff59e1038f6f3f4d4e1 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -219,4 +219,17 @@ public class EntropyConfig {
         durabilityCheckMessage = getString("entropy.unbreakable-breaker.durability-check-message", "&a[+] &7This block has %dura%/%max% durability.", "Message that shows when checking the durability of an unbreakable block.");
         unbreakableMap = UnbreakableDurability.initMap(getStringIntMap("entropy.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults(), "block: durability"));
     }
+
+    public static boolean cannonRangeLimit;
+    public static int cannonRangeLimitChunks;
+    public static boolean pearlRangeLimit;
+    public static int pearlRangeLimitChunks;
+    private static void limits() {
+        cannonRangeLimit = getBoolean("entropy.limits.cannon-range-limit", false, "Delete cannon entities if they are shot past a certain range.", "X and Z axis are separate so the allowed area has a square shape.");
+        cannonRangeLimitChunks = getInt("entropy.limits.cannon-range-limit-chunks", 40, "How many chunks cannons are allowed to shoot.");
+        pearlRangeLimit = getBoolean("entropy.limits.pearl-range-limit", false, "Remove a pearls x and z velocity if it travels a certain distance.", "Checking Radius is circular.");
+        pearlRangeLimitChunks = getInt("entropy.limits.pearl-range-limit-chunks", 10, "How many chunks are pearls allowed to travel.");
+
+        setComment("entropy.limits", "Enforces different cannoning and non cannoning related parameters.");
+    }
 }
