From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Fri, 9 Dec 2022 13:10:52 +0100
Subject: [PATCH] convert player chat to system message


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 7d6939dbe4bfda79fd9e9216748d1e5d720dd63b..1a34719de7ab303250a38c2cfbeb2ed9cc0f5ac5 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2011,10 +2011,25 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
     @Override
     public void send(Packet<?> packet) {
+        // Entropy start - disable chat message verification
+        if(EntropyConfig.disableChatMessageVerification && packet instanceof ClientboundPlayerChatPacket chat) {
+            packet = new ClientboundSystemChatPacket(chat.chatType().resolve(this.player.level.registryAccess()).get().decorate(chat.message().unsignedContent().orElseGet(() -> Component.literal(chat.message().signedContent().plain()))), false);
+            this.send(packet);
+            return;
+        }
+        // Entropy end
+
         this.send(packet, (PacketSendListener) null);
     }
 
     public void send(Packet<?> packet, @Nullable PacketSendListener callbacks) {
+        // Entropy start - disable chat message verification
+        if(EntropyConfig.disableChatMessageVerification && packet instanceof ClientboundPlayerChatPacket chat && callbacks != null) {
+            this.send(chat);
+            return;
+        }
+        // Entropy end
+
         // CraftBukkit start
         if (packet == null || this.processedDisconnect) { // Spigot
             return;
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 28f79879c5c83020493f3862381335571c2ab720..cf25f9d8c2089dc694c8bd937e76f34a414f7ae4 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1435,7 +1435,7 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(PlayerChatMessage message, ChatSender profile) { // Paper - private -> public
-        return !message.hasExpiredServer(Instant.now()) && message.verify(profile);
+        return (EntropyConfig.disableChatMessageVerification || !message.hasExpiredServer(Instant.now()) && message.verify(profile)); // Entropy - disable chat message verification
     }
 
     // CraftBukkit start
