From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Mon, 11 Dec 2023 15:04:13 +0100
Subject: [PATCH] upstream fixes


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index e7c3c0e9e7e16bab104cea7ef4830d051f623c55..5b6d4c75c77dee080d162c564a68bc881f2ce980 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1843,6 +1843,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     @Override
     public void sendBlockUpdated(BlockPos pos, BlockState oldState, BlockState newState, int flags) {
+        if(!oldState.is(newState.getBlock())) this.unbreakableDurability.tryCleanup(pos); // Entropy
         if (this.isUpdatingNavigations) {
             String s = "recursive call to sendBlockUpdated";
 
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index d138660e459b2d09ba9dd7ae2bf3da499d9ae25c..ddba74f60a9eb22c33a4c8905fd6b1ea8004c59c 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -353,6 +353,8 @@ public final class ItemStack {
             int oldCount = this.getCount();
             ServerLevel world = (ServerLevel) context.getLevel();
 
+            world.unbreakableDurability.tryCheckDurability(this, blockposition, entityhuman, world); // Entropy
+
             if (!(item instanceof BucketItem/* || item instanceof SolidBucketItem*/)) { // if not bucket // Paper - capture block states for snow buckets
                 world.captureBlockStates = true;
                 // special case bonemeal
diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 0736088e8e182100b85434d7423587dbaa8d81ec..5fa0cb1bf63bc2e9cc1dd04b45494ba3bdef7ff8 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -28,6 +28,7 @@ import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.entity.item.FallingBlockEntity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.item.PrimedTnt;
+import net.minecraft.world.entity.monster.Creeper;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.Projectile;
 import net.minecraft.world.item.ItemStack;
@@ -176,6 +177,7 @@ public class Explosion {
         // CraftBukkit end
         this.level.gameEvent(this.source, GameEvent.EXPLODE, new Vec3(this.x, this.y, this.z));
         ObjectOpenHashSet<BlockPos> blockDamageSet = new ObjectOpenHashSet<>();
+        ObjectOpenHashSet<BlockPos> unbreakableBlocks = new ObjectOpenHashSet<>();
         boolean flag = true;
 
         int i;
@@ -244,15 +246,22 @@ public class Explosion {
                                 cachedOriginResistance = optional;
                                 originCached = true;
                             }
-                            // Entropy end
 
-                            // Entropy start
                             if (!iblockdata.getFluidState().isEmpty() && (iblockdata.getBlock() instanceof SimpleWaterloggedBlock) && !EventUtil.sendCancellable(level, new ExplodeWaterloggedEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location)).isCancelled()) {
                                 blockDamageSet.add(blockposition);
                             }
                             if (iblockdata.is(Blocks.LAVA) && !EventUtil.sendCancellable(level, new ExplodeLavaEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location)).isCancelled()) {
                                 blockDamageSet.add(blockposition);
                             }
+
+                            if (EntropyConfig.unbreakableBreaker &&
+                                    !iblockdata.isAir() &&
+                                    (this.source instanceof PrimedTnt || (EntropyConfig.unbreakableCreeper && this.source instanceof Creeper)) &&
+                                    !unbreakableBlocks.contains(blockposition) &&
+                                    level.unbreakableDurability.isBreakable(blockposition, iblockdata, level)
+                                            ) {
+                                unbreakableBlocks.add(blockposition);
+                                } else
                             // Entropy end
 
                             if (f > 0.0F && this.damageCalculator.shouldBlockExplode(this, this.level, blockposition, iblockdata, f) && !iblockdata.isAir()) { // Entropy
@@ -278,7 +287,12 @@ public class Explosion {
         }
         // Entropy start
         else if (this.source instanceof PrimedTnt t) t.blockDamageSkipped = true;
-        if(this.source instanceof PrimedTnt t && (blockDamageSet.size() > 0 /*||unbreakables.size() > 0*/)) t.destroyedBlock = true;
+        if(this.source instanceof PrimedTnt t && (blockDamageSet.size() > 0 || unbreakableBlocks.size() > 0)) t.destroyedBlock = true;
+
+        for(BlockPos b : unbreakableBlocks) {
+            if(level.unbreakableDurability.reduceDurability(b)) this.toBlow.add(b);
+        }
+        // Entropy end
 
         this.toBlow.addAll(blockDamageSet);
         float f2 = this.radius * 2.0F;
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index e8765114f8d753f5295c2db3b414b69b79393cf8..79d6cb071cbcf6e10309f8d58b462b2baf75344d 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -295,6 +295,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     protected Level(WritableLevelData worlddatamutable, ResourceKey<Level> resourcekey, RegistryAccess iregistrycustom, Holder<DimensionType> holder, Supplier<ProfilerFiller> supplier, boolean flag, boolean flag1, long i, int j, org.bukkit.generator.ChunkGenerator gen, org.bukkit.generator.BiomeProvider biomeProvider, org.bukkit.World.Environment env, java.util.function.Function<org.spigotmc.SpigotWorldConfig, io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, java.util.concurrent.Executor executor) { // Paper - Async-Anti-Xray - Pass executor
         blockStateCache = EntropyConfig.cacheBlockStates ? new BlockStateCache() : new BlockStateHandler(); // Entropy
         unbreakableDurability = new UnbreakableDurability(); // Entropy
+        chunkLoader = new ChunkLoader();
 
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper
diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index fdd9c61b7248e92dbcbec91cd6fe4c6310bba237..a3e27b41fd8d43c9c8209a85870aa0b92c7199ec 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -308,6 +308,7 @@ public class Block extends BlockBehaviour implements ItemLike {
             state.spawnAfterBreak((ServerLevel) world, pos, ItemStack.EMPTY, true);
         }
 
+        world.unbreakableDurability.tryCleanup(pos);
     }
 
     public static void dropResources(BlockState state, LevelAccessor world, BlockPos pos, @Nullable BlockEntity blockEntity) {
@@ -345,6 +346,7 @@ public class Block extends BlockBehaviour implements ItemLike {
             state.spawnAfterBreak((ServerLevel) world, pos, tool, true);
         }
 
+        world.unbreakableDurability.tryCleanup(pos);
     }
 
     public static void popResource(Level world, BlockPos pos, ItemStack stack) {
diff --git a/src/main/java/xyz/fragmentmc/entropy/explosion/UnbreakableDurability.java b/src/main/java/xyz/fragmentmc/entropy/explosion/UnbreakableDurability.java
new file mode 100644
index 0000000000000000000000000000000000000000..ef4d225c677d7b6790e142e1e5840d0147509f74
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/explosion/UnbreakableDurability.java
@@ -0,0 +1,101 @@
+package xyz.fragmentmc.entropy.explosion;
+
+import it.unimi.dsi.fastutil.ints.Int2IntMap;
+import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2IntMap;
+import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.Location;
+import xyz.fragmentmc.entropy.EntropyConfig;
+import xyz.fragmentmc.entropy.event.AssignDurabilityEvent;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class UnbreakableDurability {
+    public static Map<String, Integer> getDefaults() {
+        HashMap<String, Integer> unbreakableDefaults = new HashMap<>();
+        unbreakableDefaults.put("obsidian", 5);
+        unbreakableDefaults.put("crying_obsidian", 5);
+        unbreakableDefaults.put("ancient_debris", 5);
+        unbreakableDefaults.put("anvil", 5);
+        unbreakableDefaults.put("chipped_anvil", 5);
+        unbreakableDefaults.put("damaged_anvil", 5);
+        unbreakableDefaults.put("netherite_block", 5);
+        unbreakableDefaults.put("respawn_anchor", 5);
+        unbreakableDefaults.put("enchanting_table", 5);
+        unbreakableDefaults.put("reinforced_deepslate", 5);
+        unbreakableDefaults.put("ender_chest", 5);
+        return unbreakableDefaults;
+    }
+
+    public void tryCheckDurability(ItemStack item, BlockPos b, Player player, Level l) {
+        BlockState bs = l.blockStateCache.getCached(l, b);
+        if ((item.getBukkitStack().getType().name().equalsIgnoreCase(EntropyConfig.durabilityCheckItem)) && (this.isBreakable(b, l.blockStateCache.getCached(l, b), l))) {
+            player.getBukkitEntity().sendMessage(this.displayMessage(b, bs.getBlock(), bs, l));
+        }
+    }
+
+    private final Long2IntMap unbreakableDurability = new Long2IntOpenHashMap();
+
+    public boolean isBreakable(BlockPos b, BlockState bs, Level level) {
+        if (bs.isAir()) return false;
+        if (!bs.getFluidState().isEmpty()) return false;
+        if (!unbreakableDurability.containsKey(b.asLong())) {
+            int durability = EntropyConfig.unbreakableMap.get(BuiltInRegistries.BLOCK.getKey(bs.getBlock()).getPath().hashCode());
+            if (durability == 0) return false;
+            AssignDurabilityEvent event = new AssignDurabilityEvent(durability, bs.getBukkitMaterial(), new Location(level.getWorld(), b.getX(), b.getY(), b.getZ()));
+            level.getCraftServer().getPluginManager().callEvent(event);
+            unbreakableDurability.put(b.asLong(), event.getDurability());
+        }
+        return true;
+    }
+
+    public boolean reduceDurability(BlockPos b) {
+        long key = b.asLong();
+        int durability = unbreakableDurability.get(key) - 1;
+        if (durability < 1) {
+            unbreakableDurability.remove(key);
+            return true;
+        }
+        unbreakableDurability.put(key, durability);
+        return false;
+    }
+
+    public void tryCleanup(BlockPos b) {
+        unbreakableDurability.remove(b.asLong());
+    }
+
+    public String displayMessage(BlockPos b, Block m, BlockState bs, Level l) {
+        int dura = 1;
+        int max = 1;
+        if (unbreakableDurability.containsKey(b.asLong())) {
+            dura = getDurability(b);
+            max = getMaxDurability(m, b, bs, l);
+        }
+        return EntropyConfig.durabilityCheckMessage.replaceAll("&", "§").replaceAll("%dura%", "" + dura).replaceAll("%max%", "" + max);
+    }
+
+    public int getDurability(BlockPos b) {
+        return unbreakableDurability.get(b.asLong());
+    }
+
+    public int getMaxDurability(Block b, BlockPos bp, BlockState bs, Level l) {
+        int durability = EntropyConfig.unbreakableMap.get(BuiltInRegistries.BLOCK.getKey(b).getPath().hashCode());
+        AssignDurabilityEvent event = new AssignDurabilityEvent(durability, bs.getBukkitMaterial(), new Location(l.getWorld(), bp.getX(), bp.getY(), bp.getZ()));
+        l.getCraftServer().getPluginManager().callEvent(event);
+        return event.getDurability();
+    }
+
+    public static Int2IntMap initMap(Map<String, Integer> config) {
+        Int2IntMap map = new Int2IntOpenHashMap();
+        config.forEach((key, value) -> map.put(key.hashCode(), (int) value));
+        return map;
+    }
+}
\ No newline at end of file
