From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Sat, 10 Dec 2022 13:30:48 +0100
Subject: [PATCH] use static hash values for unbreakable breaker


diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 50d7f70217c58a77ff9643c6c5deb9c770f5854f..7f45838aa89731f16b065de537831380300a67dc 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -258,7 +258,7 @@ public class EntropyConfig {
     public static boolean unbreakableCreeper;
     public static String durabilityCheckItem;
     public static String durabilityCheckMessage;
-    public static Map<String, Integer> unbreakableMap;
+    public static Map<Integer, Integer> unbreakableMap;
     private static void cannoningConfiguration() {
         disableWaterPushingTNT = getBoolean("entropy.patches.disable-water-pushing-tnt", true,
                 "Disable TNT getting pushed by water (aka. Waterflow).");
@@ -309,7 +309,7 @@ public class EntropyConfig {
                 "Item used to check an unbreakable blocks durability.");
         durabilityCheckMessage= getString("entropy.unbreakable-breaker.durability-check-message", "&a[+] &7This block has %dura%/%max% durability.",
                 "Message that shows when checking the durability of an unbreakable block.");
-        unbreakableMap= getStringIntMap("entropy.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults());
+        unbreakableMap= UnbreakableDurability.initMap(getStringIntMap("entropy.unbreakable-breaker.blocks", UnbreakableDurability.getDefaults()));
     }
 
     public static boolean disableChatMessageVerification;
diff --git a/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java b/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
index 7223d05007e39717884b3469e24214ea7f8ef9fb..fcea8377ac183f165dd84fdbd1130c5088db6de0 100644
--- a/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
+++ b/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
@@ -28,19 +28,73 @@ public class UnbreakableDurability {
         return unbreakableDefaults;
     }
 
+    public static Map<Integer, Integer> initMap(Map<String, Integer> config) {
+        HashMap<Integer, Integer> map = new HashMap<>();
+        config.forEach((key, value) -> map.put(key.hashCode(), value));
+        return map;
+    }
+
     public void tryCheckDurability(ItemStack item, BlockPos b, Player player, Level l) {
         if((item.getBukkitStack().getType().name().equalsIgnoreCase(EntropyConfig.durabilityCheckItem)) && (this.isBreakable(b, l.getBlockState(b)))) {
             player.getBukkitEntity().sendMessage(this.displayMessage(b, l.getBlockState(b).getBukkitMaterial()));
         }
     }
 
-    private Map<Long, Integer> unbreakableDurability = new HashMap<>();
+    private final Map<Long, Integer> unbreakableDurability = new HashMap<>();
 
     public boolean isBreakable(BlockPos b, BlockState bs) {
+        if(bs.isAir()) return false;
+        if(!bs.getFluidState().isEmpty()) return false;
         if(!unbreakableDurability.containsKey(b.asLong())) {
-            String material = Registry.BLOCK.getKey(bs.getBlock()).getPath();
-            if(EntropyConfig.unbreakableMap.containsKey(material)) unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(material));
-            else return false;
+            switch (Registry.BLOCK.getKey(bs.getBlock()).getPath()) {
+                case "obsidian" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(351871579));
+                    return true;
+                }
+                case "crying_obsidian" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-1768395518));
+                    return true;
+                }
+                case "ancient_debris" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-1760410398));
+                    return true;
+                }
+                case "anvil" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(92975308));
+                    return true;
+                }
+                case "chipped_anvil" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(2049113808));
+                    return true;
+                }
+                case "damaged_anvil" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-5782814));
+                    return true;
+                }
+                case "netherite_block" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-605774000));
+                    return true;
+                }
+                case "respawn_anchor" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-1293396564));
+                    return true;
+                }
+                case "enchanting_table" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-758067976));
+                    return true;
+                }
+                case "reinforced_deepslate" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-1618455329));
+                    return true;
+                }
+                case "ender_chest" -> {
+                    unbreakableDurability.put(b.asLong(), EntropyConfig.unbreakableMap.get(-761143606));
+                    return true;
+                }
+                default -> {
+                    return false;
+                }
+            }
         }
         return true;
     }
