From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Sat, 10 Dec 2022 14:36:41 +0100
Subject: [PATCH] cache explosion damage


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 59726d579bdefdb77855bc0ec23325cd352e48e2..1640eb2a451e8be7bba54eaa6b8be725919935f7 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -43,6 +43,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.block.BlockExplodeEvent;
 import org.bukkit.event.entity.EntityExplodeEvent;
+import xyz.fragmentmc.entropy.CachedEntityExplosionDamageCalculator;
 import xyz.fragmentmc.entropy.EntropyConfig;
 import xyz.fragmentmc.entropy.ExplosionProofItems;
 
@@ -104,7 +105,7 @@ public class Explosion {
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(@Nullable Entity entity) {
-        return (ExplosionDamageCalculator) (entity == null ? Explosion.EXPLOSION_DAMAGE_CALCULATOR : new EntityBasedExplosionDamageCalculator(entity));
+        return entity == null ? Explosion.EXPLOSION_DAMAGE_CALCULATOR : EntropyConfig.cacheExplosionDamage ? new CachedEntityExplosionDamageCalculator(entity) : new EntityBasedExplosionDamageCalculator(entity); // Entropy - cache explosion damage
     }
 
     public static float getSeenPercent(Vec3 source, Entity entity) {
diff --git a/src/main/java/xyz/fragmentmc/entropy/CachedEntityExplosionDamageCalculator.java b/src/main/java/xyz/fragmentmc/entropy/CachedEntityExplosionDamageCalculator.java
new file mode 100644
index 0000000000000000000000000000000000000000..f115c016c2fdd60fcff3ade3fcf9c2385b2c46f7
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/CachedEntityExplosionDamageCalculator.java
@@ -0,0 +1,47 @@
+package xyz.fragmentmc.entropy;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.item.PrimedTnt;
+import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.Explosion;
+import net.minecraft.world.level.ExplosionDamageCalculator;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.material.FluidState;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Optional;
+
+public class CachedEntityExplosionDamageCalculator extends ExplosionDamageCalculator {
+    private final Entity source;
+    private final boolean isTnt;
+    private static final Int2ObjectMap<Optional<Float>> cache = new Int2ObjectOpenHashMap<>();
+
+    public CachedEntityExplosionDamageCalculator(Entity entity) {
+        this.source = entity;
+        this.isTnt = entity instanceof PrimedTnt;
+    }
+
+    @Override
+    public @NotNull Optional<Float> getBlockExplosionResistance(@NotNull Explosion explosion, @NotNull BlockGetter world, @NotNull BlockPos pos, @NotNull BlockState blockState, @NotNull FluidState fluidState) {
+        if(isTnt) {
+            if(blockState.isAir() && fluidState.isEmpty()) return Optional.empty();
+            final int key = blockState.getBlock().hashCode() ^ blockState.getFluidState().getType().hashCode();
+            Optional<Float> resistance = cache.get(key);
+            if(resistance == null) {
+                resistance = super.getBlockExplosionResistance(explosion, world, pos, blockState, fluidState);
+                cache.put(key, resistance);
+            }
+            return resistance;
+        }
+        return super.getBlockExplosionResistance(explosion, world, pos, blockState, fluidState).map((max) -> this.source.getBlockExplosionResistance(explosion, world, pos, blockState, fluidState, max));
+    }
+
+    @Override
+    public boolean shouldBlockExplode(@NotNull Explosion explosion, @NotNull BlockGetter world, @NotNull BlockPos pos, @NotNull BlockState state, float power) {
+        if(isTnt) return true;
+        return this.source.shouldBlockExplode(explosion, world, pos, state, power);
+    }
+}
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 7f45838aa89731f16b065de537831380300a67dc..bd0a52114ccaa2d32c6388fb0447e1c81c70c38f 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -248,6 +248,7 @@ public class EntropyConfig {
     public static boolean combinedExplosions;
     public static boolean explosionWaterCheck;
     public static boolean skipRayOrigin;
+    public static boolean cacheExplosionDamage;
     public static boolean tntLoadChunks;
     public static boolean fallingBlockLoadChunks;
     public static boolean enderPearlLoadChunks;
@@ -283,6 +284,8 @@ public class EntropyConfig {
                 "Disable block breaking checks when tnt explodes while submerged.");
         skipRayOrigin= getBoolean("entropy.optimizations.skip-ray-origin", true,
                 "Skip the origin of explosion rays, as it gets checked with every ray.");
+        cacheExplosionDamage= getBoolean("entropy.optimizations.cache-explosion-damage", true,
+                "Cache explosion damage values");
         setComment("entropy.optimizations", "Changes that improve cannoning performance");
 
         tntLoadChunks= getBoolean("entropy.chunks.tnt-loads-chunks", true,
