From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Mon, 11 Dec 2023 15:53:18 +0100
Subject: [PATCH] tnt flow packet limiter


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 2843837c419176bf00f0fd2ccb23e5f14e60ff85..ffa880984eaaf20b536540573481246d527291a1 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -146,7 +146,8 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         }
 
         // Paper start - Optional prevent TNT from moving in water
-        if (!this.isRemoved() && this.wasTouchingWater && EventUtil.sendCancellable(level, new TntFlowEvent(EntropyConfig.disableTNTFlow, new CraftEntropyTntPrimed(level.getCraftServer(), this))).isCancelled()) { // Entropy
+        boolean doUpdate = !EntropyConfig.packetLimiterTntFlow || this.getFuse() % 5 == 0; // Entropy
+        if (!this.isRemoved() && this.wasTouchingWater && doUpdate && EventUtil.sendCancellable(level, new TntFlowEvent(EntropyConfig.disableTNTFlow, new CraftEntropyTntPrimed(level.getCraftServer(), this))).isCancelled()) { // Entropy
             /*
              * Author: Jedediah Smith <jedediah@silencegreys.com>
              */
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 70cdd642b880c6d3cc2465e96d849c27a7767687..0fc09c7067341b17641ccd1356bdfedd0b25eab7 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -52,7 +52,7 @@ public class EntropyConfig {
                 throw new IOException(e);
             }
         }
-        getString("info.version", "1.3");
+        getString("info.version", "1.4");
         setComment("info",
                 "====================================",
                 "Entropy Cannoning Jar by FragmentMC",
@@ -208,6 +208,13 @@ public class EntropyConfig {
         setComment("entropy.optimizations", "Changes that improve cannoning performance.", "* - slight behaviour changes, should not be noticeable.");
     }
 
+    public static boolean packetLimiterTntFlow;
+    private static void packetLimiter() {
+        packetLimiterTntFlow = getBoolean("entropy.packet-limiter.tnt-flow", true, "Update tnt flowing in water less frequently.");
+
+        setComment("entropy.packet-limiter", "Limit packets being sent to the client for performance reasons.");
+    }
+
     public static boolean tntLoadsChunks;
     public static boolean fallingBlockLoadsChunks;
     public static boolean enderpearlLoadsChunks;
