From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Sat, 10 Dec 2022 15:51:29 +0100
Subject: [PATCH] replace hash set


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 04ae1e7c7e557f986aa34b73859f82291b9e9153..aa2d3cb7b58fe817c5a9fe6867881bfd4c64752c 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -1,10 +1,10 @@
 package net.minecraft.world.level;
 
 import com.google.common.collect.Maps;
-import com.google.common.collect.Sets;
 import com.mojang.datafixers.util.Pair;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectListIterator;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
@@ -46,7 +46,9 @@ import org.bukkit.event.entity.EntityExplodeEvent;
 import xyz.fragmentmc.entropy.*;
 
 import javax.annotation.Nullable;
-import java.util.*;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 // CraftBukkit end
 
 public class Explosion {
@@ -164,13 +166,12 @@ public class Explosion {
         }
         // CraftBukkit end
         this.level.gameEvent(this.source, GameEvent.EXPLODE, new Vec3(this.x, this.y, this.z));
-        Set<BlockPos> set = Sets.newHashSet();
-        boolean flag = true;
+        ObjectOpenHashSet<BlockPos> blockDamageSet = new ObjectOpenHashSet<>(); // Entropy - replace hash set
 
         int i;
         int j;
 
-        HashSet<BlockPos> unbreakables = Sets.newHashSet();  // Entropy - unbreakable breaker
+        ObjectOpenHashSet<BlockPos> unbreakables = new ObjectOpenHashSet<>();  // Entropy - unbreakable breaker
 
         // Entropy start - skip ray origin
         boolean originCached = false;
@@ -239,10 +240,10 @@ public class Explosion {
 
                             // Entropy start - explode waterlogged blocks and lava
                             if (EntropyConfig.explodeWaterloggedBlocks && !iblockdata.getFluidState().isEmpty() && !iblockdata.getMaterial().isLiquid()) {
-                                set.add(blockposition);
+                                blockDamageSet.add(blockposition);
                             }
                             if (EntropyConfig.explodeLava && iblockdata.getMaterial() == Material.LAVA) {
-                                set.add(blockposition);
+                                blockDamageSet.add(blockposition);
                             }
                             // Entropy end
 
@@ -256,13 +257,13 @@ public class Explosion {
                             }
                             else // Entropy end
                             if (f > 0.0F && this.damageCalculator.shouldBlockExplode(this, this.level, blockposition, iblockdata, f)) {
-                                set.add(blockposition);
+                                blockDamageSet.add(blockposition); // Entropy - replace hash set
                                 // Paper start - prevent headless pistons from forming
                                 if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowHeadlessPistons && iblockdata.getBlock() == Blocks.MOVING_PISTON) {
                                     BlockEntity extension = this.level.getBlockEntity(blockposition);
                                     if (extension instanceof net.minecraft.world.level.block.piston.PistonMovingBlockEntity blockEntity && blockEntity.isSourcePiston()) {
                                        net.minecraft.core.Direction direction = iblockdata.getValue(net.minecraft.world.level.block.piston.PistonHeadBlock.FACING);
-                                       set.add(blockposition.relative(direction.getOpposite()));
+                                       blockDamageSet.add(blockposition.relative(direction.getOpposite())); // Entropy - replace hash set
                                     }
                                 }
                                 // Paper end
@@ -278,11 +279,11 @@ public class Explosion {
         }
         // Entropy start - unbreakable breaker
         for(BlockPos b : unbreakables) {
-            if (level.unbreakableDurability.reduceDurability(b)) set.add(b);
+            if (level.unbreakableDurability.reduceDurability(b)) blockDamageSet.add(b);
         }
         // Entropy end
 
-        this.toBlow.addAll(set);
+        this.toBlow.addAll(blockDamageSet);
         float f2 = this.radius * 2.0F;
 
         i = Mth.floor(this.x - (double) f2 - 1.0D);
