From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Mon, 11 Dec 2023 18:02:18 +0100
Subject: [PATCH] fix api imp and non full block checks


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 2b38bd00b7c046600ef6fa3f346814bf5fcedb49..8dbf7c5d9e8aa5704c330680513d427496617c02 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1174,7 +1174,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
                 // Entropy start
                 if((this instanceof PrimedTnt || this instanceof FallingBlockEntity) && EventUtil.sendCancellable(level, new NonFullBlockCollideEvent(EntropyConfig.disableNonFullBlockHitbox, this.getBukkitEntity(), new Location(level.getWorld(), this.getX(), this.getY(), this.getZ()))).isCancelled() && !CannonUtil.isInCannon(this)) {
-                    BlockPos inside = new BlockPos((int) this.getX(), (int) this.getY(), (int) this.getZ());;
+                    BlockPos inside = this.blockPosition;
                     if(movement.x > vec3d1.x) {
                         CannonUtil.checkNonFullCollision(inside.east(), this, this.level);
                     }
diff --git a/src/main/java/xyz/fragmentmc/entropy/util/CannonUtil.java b/src/main/java/xyz/fragmentmc/entropy/util/CannonUtil.java
index d760f89251a9e01ea8cc38f064f696279d9d94de..3adc827e672ee0e74b511dbdfc9aae53233b62eb 100644
--- a/src/main/java/xyz/fragmentmc/entropy/util/CannonUtil.java
+++ b/src/main/java/xyz/fragmentmc/entropy/util/CannonUtil.java
@@ -6,6 +6,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.event.Cancellable;
 import org.bukkit.util.Vector;
 import xyz.fragmentmc.entropy.event.IsInCannonEvent;
 
@@ -22,10 +23,11 @@ public class CannonUtil {
         double z = originVector.getZ() - entity.getZ();
         boolean isInCannon = x * x + z * z < 48 * 48;
         org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity();
-        IsInCannonEvent event = new IsInCannonEvent(isInCannon, bukkitEntity, bukkitEntity.getLocation());
-        return event.callEvent();
+        Cancellable event = EventUtil.sendCancellable(new IsInCannonEvent(isInCannon, bukkitEntity, bukkitEntity.getLocation()));
+        return event.isCancelled();
     }
     public static void checkNonFullCollision(BlockPos relevant, Entity entity, Level level) {
+        if(level.blockStateCache.getCached(level, relevant).isAir()) return;
         if (CollisionUtil.voxelShapeIntersect(entity.getBoundingBox(), new AABB(relevant))) {
             level.setBlockAndUpdate(relevant, Blocks.COBBLESTONE.defaultBlockState());
         }
diff --git a/src/main/java/xyz/fragmentmc/entropy/util/EventUtil.java b/src/main/java/xyz/fragmentmc/entropy/util/EventUtil.java
index 026e012149255b3e41963687364c3241a05d3e86..0b1f06b53df47399cb05a4cccb79909672aa6de3 100644
--- a/src/main/java/xyz/fragmentmc/entropy/util/EventUtil.java
+++ b/src/main/java/xyz/fragmentmc/entropy/util/EventUtil.java
@@ -1,6 +1,7 @@
 package xyz.fragmentmc.entropy.util;
 
 import net.minecraft.world.level.Level;
+import org.bukkit.Bukkit;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.Event;
 
@@ -9,4 +10,9 @@ public class EventUtil {
         level.getCraftServer().getPluginManager().callEvent(event);
         return (Cancellable) event;
     }
+
+    public static Cancellable sendCancellable(Event event) {
+        Bukkit.getPluginManager().callEvent(event);
+        return (Cancellable) event;
+    }
 }
