From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Sat, 10 Dec 2022 17:16:39 +0100
Subject: [PATCH] reduce rays


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index aa2d3cb7b58fe817c5a9fe6867881bfd4c64752c..d952c181d03f7778da02630cb4f8d7f7569eedef 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -184,14 +184,17 @@ public class Explosion {
             BlockState blockState = this.level.getBlockState(this.source.blockPosition);
             doBlockBreakCheck=!(EntropyConfig.explodeLava ? blockState.getMaterial().equals(Material.WATER) : blockState.getMaterial().isLiquid());
         }
+        int projectionSize = EntropyConfig.reduceRays ? EntropyConfig.rayProjectionSize : 15; // Entropy - reduce rays
         if(doBlockBreakCheck) // Entropy end
-        for (int k = 0; k < 16; ++k) {
-            for (i = 0; i < 16; ++i) {
-                for (j = 0; j < 16; ++j) {
-                    if (k == 0 || k == 15 || i == 0 || i == 15 || j == 0 || j == 15) {
-                        double d0 = (double) ((float) k / 15.0F * 2.0F - 1.0F);
-                        double d1 = (double) ((float) i / 15.0F * 2.0F - 1.0F);
-                        double d2 = (double) ((float) j / 15.0F * 2.0F - 1.0F);
+        for (int k = 0; k < projectionSize+1; ++k) {  // Entropy - reduce rays
+            for (i = 0; i < projectionSize+1; ++i) {  // Entropy - reduce rays
+                for (j = 0; j < projectionSize+1; ++j) {  // Entropy - reduce rays
+                    // Entropy start - reduce rays
+                    if ((k == 0 || k == projectionSize || i == 0 || i == projectionSize || j == 0 || j == projectionSize)) {
+                        double d0 = ((float) k / projectionSize * 2.0F - 1.0F);
+                        double d1 = ((float) i / projectionSize * 2.0F - 1.0F);
+                        double d2 = ((float) j / projectionSize* 2.0F - 1.0F);
+                        // Entropy end
                         double d3 = Math.sqrt(d0 * d0 + d1 * d1 + d2 * d2);
 
                         d0 /= d3;
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 03b8a0fa50352a8e37abe0e426490c498a3b3a50..eb58c6bf4fee642fbcbcedda36325ff70c8aa10b 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -251,6 +251,8 @@ public class EntropyConfig {
     public static boolean cacheExplosionDamage;
     public static boolean cacheBlockDensity;
     public static boolean staticRayStrenght;
+    public static boolean reduceRays;
+    public static int rayProjectionSize;
     public static boolean tntLoadChunks;
     public static boolean fallingBlockLoadChunks;
     public static boolean enderPearlLoadChunks;
@@ -291,8 +293,13 @@ public class EntropyConfig {
         cacheBlockDensity= getBoolean("entropy.optimizations.cache-block-density", true,
                 "Cache block density for each explosion.");
         staticRayStrenght= getBoolean("entropy.optimizations.static-ray-strength", true,
-                "Remove the random component in explosion damage, CAN change behavior slightly but should not be noticeable.");
-        setComment("entropy.optimizations", "Changes that improve cannoning performance");
+                "Remove the random component in explosion damage. *");
+        reduceRays= getBoolean("entropy.optimizations.reduce-rays.enabled", true,
+                "Reduces the amount of rays for each explosion. *");
+        rayProjectionSize= getInt("entropy.optimizations.reduce-rays.projection-size", 11,
+                "Size of the projection cube for explosion rays (vanilla is 15), amount of rays = (n^2)*6.");
+        setComment("entropy.optimizations", "Changes that improve cannoning performance",
+                "* - slight behaviour changes, should not be noticeable");
 
         tntLoadChunks= getBoolean("entropy.chunks.tnt-loads-chunks", true,
         "If TNT tries to travel through an unloaded chunk that chunk will be loaded.");
