From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Mon, 12 Dec 2022 11:05:36 +0100
Subject: [PATCH] fixed config comments


diff --git a/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java b/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
index fcea8377ac183f165dd84fdbd1130c5088db6de0..1efd410355da099fb3e4ca6e05210b56f8681f6f 100644
--- a/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
+++ b/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
@@ -129,6 +129,6 @@ public class UnbreakableDurability {
     }
 
     public int getMaxDurability(Material m) {
-        return EntropyConfig.unbreakableMap.get(m.name().toLowerCase());
+        return EntropyConfig.unbreakableMap.get(m.name().toLowerCase().hashCode());
     }
 }
