From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Wed, 3 Jul 2024 15:10:01 +0200
Subject: [PATCH] fix origin location


diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 4b5ad38524296cb5cf90ee4844377e3207b3d8c1..f0dce3f415b8a8c14c43fac0f63c51426edcb4da 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -47,6 +47,7 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
 import org.bukkit.util.Vector;
 import org.slf4j.Logger;
 
@@ -92,6 +93,9 @@ public class FallingBlockEntity extends Entity {
         this.yo = y;
         this.zo = z;
         this.setStartPos(this.blockPosition());
+
+        // Entropy Start - entity origin fixes
+        this.setOrigin(new Location(world.getWorld(), x, y, z));
     }
 
     public static FallingBlockEntity fall(Level world, BlockPos pos, BlockState state) {
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index ffa880984eaaf20b536540573481246d527291a1..81d74571c9414f37b4f41b8a0eee375682eafa92 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -15,6 +15,7 @@ import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.TraceableEntity;
 import net.minecraft.world.level.Level;
 // CraftBukkit start;
+import org.bukkit.Location;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
 import org.bukkit.util.Vector;
@@ -59,6 +60,9 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         this.yo = y;
         this.zo = z;
         this.owner = igniter;
+
+        // Entropy Start - entity origin fixes
+        this.setOrigin(new Location(world.getWorld(), x, y, z));
     }
 
     @Override
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 5f3ac5edb9b6601a2df8f0f3c9b0a4f1e71f8caa..bd60096c6a8674701de03524fb5fb03ab28e4654 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -200,7 +200,7 @@ public class EntropyConfig {
         cacheExplosionDamage = getBoolean("entropy.optimizations.cache-explosion-damage", true, "Cache explosion damage values.");
         combinedExplosions = getBoolean("entropy.optimizations.combined-explosions", true, "Do one powerful explosion instead of unmerging TNT, when possible.");
         staticRayStrength = getBoolean("entropy.optimizations.static-ray-strength", true, "* Remove the random component in explosion damage.");
-        cacheBlockDensity = getBoolean("entropy.optimizations.cache-block-density", true, "Cache block density for explosions.");
+        cacheBlockDensity = getBoolean("entropy.optimizations.cache-block-density", true, "* Cache block density for explosions.");
         reduceRays = getBoolean("entropy.optimizations.reduce-rays.enabled", true, "* Reduces the amount of rays for explosion block damage.");
         rayProjectionSize = getInt("entropy.optimizations.reduce-rays.projection-size", 11, "Size of the projection cube for explosion rays (vanilla is 15), amount of rays = (n^2)*6.");
         cacheBlockStates = getBoolean("entropy.optimizations.cache-block-states", true, "Cache block states in between explosions when possible.");
