From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Wed, 3 Jul 2024 16:11:30 +0200
Subject: [PATCH] ip hashing


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 52cb6476604c98465f81ff9e8d8344d7536a8868..911021190227e58bb4418bb7d4424229f789263f 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -47,6 +47,7 @@ import org.apache.commons.lang3.Validate;
 import org.slf4j.Logger;
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
+import xyz.fragmentmc.entropy.connection.IPMap;
 
 public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
@@ -182,7 +183,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public void channelActive(ChannelHandlerContext channelhandlercontext) throws Exception {
         super.channelActive(channelhandlercontext);
         this.channel = channelhandlercontext.channel();
-        this.address = this.channel.remoteAddress();
+        this.address = IPMap.Instance.getAddress((InetSocketAddress) this.channel.remoteAddress()); // Entropy - hashed ips
         // Spigot Start
         this.preparing = false;
         // Spigot End
diff --git a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
index 44d99e89226adb6234b9405f25ac9dab9bd84297..459528a8cbdf67be6066bfe9907472781623b892 100644
--- a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
+++ b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
@@ -28,6 +28,7 @@ import io.netty.util.Timeout;
 import io.netty.util.Timer;
 import java.io.IOException;
 import java.net.InetAddress;
+import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.util.Collections;
 import java.util.Iterator;
@@ -46,6 +47,7 @@ import net.minecraft.network.protocol.game.ClientboundDisconnectPacket;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.LazyLoadedValue;
 import org.slf4j.Logger;
+import xyz.fragmentmc.entropy.connection.IPMap;
 
 public class ServerConnectionListener {
 
@@ -143,7 +145,7 @@ public class ServerConnectionListener {
                                         SocketAddress socketaddr = new java.net.InetSocketAddress(realaddress, realport);
 
                                         Connection connection = (Connection) channel.pipeline().get("packet_handler");
-                                        connection.address = socketaddr;
+                                        connection.address = IPMap.Instance.getAddress((InetSocketAddress) socketaddr); // Entropy - hashed ips
                                     }
                                 } else {
                                     super.channelRead(ctx, msg);
diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 63cf71940f6480c593a43bd39900c50676367404..4f0d09b13c7da8a3f54a1920e858adc101270280 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -10,9 +10,11 @@ import net.minecraft.network.protocol.handshake.ServerHandshakePacketListener;
 import net.minecraft.network.protocol.login.ClientboundLoginDisconnectPacket;
 import net.minecraft.network.protocol.status.ServerStatus;
 import net.minecraft.server.MinecraftServer;
+import xyz.fragmentmc.entropy.connection.IPMap;
 
 // CraftBukkit start
 import java.net.InetAddress;
+import java.net.InetSocketAddress;
 import java.util.HashMap;
 // CraftBukkit end
 
@@ -48,7 +50,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                     if (!(this.connection.channel.localAddress() instanceof io.netty.channel.unix.DomainSocketAddress)) { // Paper - the connection throttle is useless when you have a Unix domain socket
                     long currentTime = System.currentTimeMillis();
                     long connectionThrottle = this.server.server.getConnectionThrottle();
-                    InetAddress address = ((java.net.InetSocketAddress) this.connection.getRemoteAddress()).getAddress();
+                    InetAddress address = (IPMap.Instance.revertAddress((InetSocketAddress) this.connection.getRemoteAddress())).getAddress(); // Entropy - hashed ips
 
                     synchronized (ServerHandshakePacketListenerImpl.throttleTracker) {
                         if (ServerHandshakePacketListenerImpl.throttleTracker.containsKey(address) && !"127.0.0.1".equals(address.getHostAddress()) && currentTime - ServerHandshakePacketListenerImpl.throttleTracker.get(address) < connectionThrottle) {
@@ -97,7 +99,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                 boolean handledByEvent = false;
                 // Try and handle the handshake through the event
                 if (com.destroystokyo.paper.event.player.PlayerHandshakeEvent.getHandlerList().getRegisteredListeners().length != 0) { // Hello? Can you hear me?
-                    java.net.SocketAddress socketAddress = this.connection.address;
+                    java.net.SocketAddress socketAddress = IPMap.Instance.revertAddress((InetSocketAddress) this.connection.address); // Entropy - hashed ips
                     String hostnameOfRemote = socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getHostString() : InetAddress.getLoopbackAddress().getHostAddress();
                     com.destroystokyo.paper.event.player.PlayerHandshakeEvent event = new com.destroystokyo.paper.event.player.PlayerHandshakeEvent(packet.hostName, hostnameOfRemote, !proxyLogicEnabled);
                     if (event.callEvent()) {
@@ -110,7 +112,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                         }
 
                         if (event.getServerHostname() != null) packet.hostName = event.getServerHostname();
-                        if (event.getSocketAddressHostname() != null) this.connection.address = new java.net.InetSocketAddress(event.getSocketAddressHostname(), socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getPort() : 0);
+                        if (event.getSocketAddressHostname() != null) this.connection.address = IPMap.Instance.getAddress(new java.net.InetSocketAddress(event.getSocketAddressHostname(), socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getPort() : 0)); // Entropy - hashed ips
                         this.connection.spoofedUUID = event.getUniqueId();
                         this.connection.spoofedProfile = gson.fromJson(event.getPropertiesJson(), com.mojang.authlib.properties.Property[].class);
                         handledByEvent = true; // Hooray, we did it!
@@ -124,9 +126,9 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                     // if (org.spigotmc.SpigotConfig.bungee) { // Paper - comment out, we check above!
                         if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.BYPASS_HOSTCHECK || ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) { // Paper
                             // Paper start - Unix domain socket support
-                            java.net.SocketAddress socketAddress = connection.getRemoteAddress();
+                            java.net.SocketAddress socketAddress = IPMap.Instance.revertAddress((InetSocketAddress) connection.getRemoteAddress()); // Entropy - hashed ips
                             packet.hostName = split[0];
-                            connection.address = new java.net.InetSocketAddress(split[1], socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getPort() : 0);
+                            connection.address = IPMap.Instance.getAddress(new java.net.InetSocketAddress(split[1], socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getPort() : 0)); // Entropy - hashed ips
                             // Paper end
                             connection.spoofedUUID = com.mojang.util.UUIDTypeAdapter.fromString( split[2] );
                         } else
diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 878001928327d92423d5f7f6d5ce8772d6fa477f..8392b9a414671782866d97bea7b87692c047eb2e 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -42,6 +42,7 @@ import org.slf4j.Logger;
 import org.bukkit.craftbukkit.util.Waitable;
 import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
 import org.bukkit.event.player.PlayerPreLoginEvent;
+import xyz.fragmentmc.entropy.connection.IPMap;
 // CraftBukkit end
 
 public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener, TickablePacketListener {
@@ -363,7 +364,7 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
 
             @Nullable
             private InetAddress getAddress() {
-                SocketAddress socketaddress = ServerLoginPacketListenerImpl.this.connection.getRemoteAddress();
+                SocketAddress socketaddress = IPMap.Instance.revertAddress((InetSocketAddress) ServerLoginPacketListenerImpl.this.connection.getRemoteAddress()); // Entropy - hashed ips
 
                 return ServerLoginPacketListenerImpl.this.server.getPreventProxyConnections() && socketaddress instanceof InetSocketAddress ? ((InetSocketAddress) socketaddress).getAddress() : null;
             }
@@ -447,12 +448,12 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
                 throw new IllegalStateException("Unsupported forwarding version " + version + ", wanted upto " + com.destroystokyo.paper.proxy.VelocityProxy.MAX_SUPPORTED_FORWARDING_VERSION);
             }
 
-            java.net.SocketAddress listening = this.connection.getRemoteAddress();
+            java.net.SocketAddress listening = IPMap.Instance.revertAddress((InetSocketAddress) this.connection.getRemoteAddress()); // Entropy - hashed ips
             int port = 0;
             if (listening instanceof java.net.InetSocketAddress) {
                 port = ((java.net.InetSocketAddress) listening).getPort();
             }
-            this.connection.address = new java.net.InetSocketAddress(com.destroystokyo.paper.proxy.VelocityProxy.readAddress(buf), port);
+            this.connection.address = IPMap.Instance.getAddress(new java.net.InetSocketAddress(com.destroystokyo.paper.proxy.VelocityProxy.readAddress(buf), port)); // Entropy - hashed ips
 
             this.gameProfile = com.destroystokyo.paper.proxy.VelocityProxy.createProfile(buf);
 
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 0e05c5aa227910b8f8d0e363c19fc0952e36d46c..dbafd90d4c1238ce4fb9e843de617cd8e77507bc 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1330,7 +1330,7 @@ public abstract class Player extends LivingEntity {
                                 long delay = currentMillis - lastSprintKnockback;
                                 if (flag1) {
                                     float knockback = 0.0f;
-                                    if (delay >= EntropyConfig.knockbackDelay) {
+                                    if (delay >= EntropyConfig.sprintingKnockbackDelay) {
                                         knockback = (float)EntropyConfig.sprintingKnockback;
                                         this.lastSprintKnockback = currentMillis;
                                     }
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index bd60096c6a8674701de03524fb5fb03ab28e4654..50fbc528da59f7d34cebf130ae94e5e9272cab7a 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -303,5 +303,16 @@ public class EntropyConfig {
         knockbackResistance = getBoolean("entropy.combat.knockback-resistance", true, "Toggle knockback resistance.");
         sprintingKnockback = getDouble("entropy.combat.sprinting-knockback", 0.5, "Add extra sprinting knockback.");
         sprintingKnockbackDelay = getInt("entropy.combat.sprinting-knockback-delay", 0, "Delay for sprinting knockback.");
+
+        setComment("entropy.combat", "Changes to pvp and pve combat.");
+    }
+
+    public static boolean ipHashing;
+    public static String ipHashingSalt;
+    private static void networking() {
+        ipHashing = getBoolean("entropy.networking.ip-hasing", false, "Replace visible ip addresses with hashes to protect user data.");
+        ipHashingSalt = getString("entropy.networking.ip-hashing-salt", "replace-this", "Set this value to a string that is not used anywhere else. Changing this value will change hashed ips.");
+
+        setComment("entropy.networking", "Changes to the network stack.");
     }
 }
diff --git a/src/main/java/xyz/fragmentmc/entropy/connection/IPMap.java b/src/main/java/xyz/fragmentmc/entropy/connection/IPMap.java
new file mode 100644
index 0000000000000000000000000000000000000000..51e19cdb445ddff9240efe1edbf55828532be8ab
--- /dev/null
+++ b/src/main/java/xyz/fragmentmc/entropy/connection/IPMap.java
@@ -0,0 +1,65 @@
+package xyz.fragmentmc.entropy.connection;
+
+import com.google.common.hash.Hashing;
+import xyz.fragmentmc.entropy.EntropyConfig;
+
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+
+public class IPMap {
+    public static IPMap Instance = new IPMap();
+
+    private HashMap<InetSocketAddress, InetSocketAddress> RealToFakeIp = new HashMap<>();
+    private HashMap<InetSocketAddress, InetSocketAddress> FakeToRealIp = new HashMap<>();
+
+    private void addIp(InetSocketAddress real, InetSocketAddress fake) {
+        RealToFakeIp.put(real, fake);
+        FakeToRealIp.put(fake, real);
+    }
+
+    private InetSocketAddress getReal(InetSocketAddress fake) {
+        return FakeToRealIp.get(fake);
+    }
+
+    private InetSocketAddress getFake(InetSocketAddress real) {
+        return RealToFakeIp.get(real);
+    }
+
+    private InetSocketAddress createFake(InetSocketAddress real) {
+        byte[] ipBytes = real.getAddress().getAddress();
+        String salt = EntropyConfig.ipHashingSalt;
+        int length = ipBytes.length + salt.length();
+        byte[] bytes = new byte[length];
+        for (int i = 0; i < length; i++) {
+            if (i < ipBytes.length) bytes[i] = ipBytes[i];
+            else bytes[i] = (byte)salt.charAt(i - ipBytes.length);
+        }
+        int hash = Hashing.sha256().hashBytes(bytes).asInt();
+        InetAddress fake;
+        try {
+            fake = Inet4Address.getByAddress(new byte[] {(byte)hash, (byte) (hash >> 8), (byte) (hash >> 16), (byte) (hash >> 24) });
+        } catch (UnknownHostException e) {
+            throw new RuntimeException(e);
+        }
+        return new InetSocketAddress(fake, 0);
+    }
+
+    public InetSocketAddress getAddress(InetSocketAddress real) {
+        if (!EntropyConfig.ipHashing) return real;
+        InetSocketAddress fake = getFake(real);
+        if (fake == null) {
+            fake = createFake(real);
+            addIp(real, fake);
+        }
+        return fake;
+    }
+
+    public InetSocketAddress revertAddress(InetSocketAddress fake) {
+        if (!EntropyConfig.ipHashing) return fake;
+        InetSocketAddress real = getReal(fake);
+        return real == null ? fake : real;
+    }
+}
