From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Mon, 12 Dec 2022 21:24:58 +0100
Subject: [PATCH] fix ender pearl chunk loading


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c6bd406ea54e78397c699fd216d997c2425793ed..bbca3a09bdc56f7c9f525e2921ac1a1aa900d19d 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -50,7 +50,6 @@ import net.minecraft.world.entity.item.FallingBlockEntity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.item.PrimedTnt;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.entity.projectile.ThrownEnderpearl;
 import net.minecraft.world.entity.vehicle.Boat;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
@@ -1350,7 +1349,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             }
 
             // Entropy start - tnt and falling block chunk loading
-            boolean load = (EntropyConfig.tntLoadChunks && this instanceof PrimedTnt) || (EntropyConfig.fallingBlockLoadChunks && this instanceof FallingBlockEntity) || (EntropyConfig.enderPearlLoadChunks && this instanceof ThrownEnderpearl);
+            boolean load = (EntropyConfig.tntLoadChunks && this instanceof PrimedTnt) || (EntropyConfig.fallingBlockLoadChunks && this instanceof FallingBlockEntity);
             io.papermc.paper.util.CollisionUtil.getCollisions(world, this, collisionBox, potentialCollisions, load, this.level.paperConfig().chunks.preventMovingIntoUnloadedChunks,
                 false, false, null, null);
             // Entropy end
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrowableProjectile.java b/src/main/java/net/minecraft/world/entity/projectile/ThrowableProjectile.java
index 94d09b05737679b133ec462815b010b19c01b4fa..91964452bee56899b5db811ff0d1bdeef1cf9ab6 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -2,8 +2,12 @@ package net.minecraft.world.entity.projectile;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.TicketType;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BlockEntity;
@@ -13,6 +17,7 @@ import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import xyz.fragmentmc.entropy.EntropyConfig;
 
 public abstract class ThrowableProjectile extends Projectile {
 
@@ -108,6 +113,16 @@ public abstract class ThrowableProjectile extends Projectile {
             this.setDeltaMovement(vec3d1.x, vec3d1.y - (double) this.getGravity(), vec3d1.z);
         }
 
+        // Entropy start - ender pearl chunk loading
+        if (EntropyConfig.enderPearlLoadChunks && this instanceof ThrownEnderpearl) {
+            Vec3 pos = this.position();
+            Vec3 vel = this.getDeltaMovement();
+            double nextX = pos.x + vel.x;
+            double nextZ = pos.z + vel.z;
+            ChunkPos chunk = new ChunkPos((int) Math.floor(nextX) >> 4, (int) Math.floor(nextZ) >> 4);
+            ((ServerLevel) level).getChunkSource().chunkMap.getDistanceManager().addTicket(TicketType.ENTROPY_LOAD, chunk, 31, MinecraftServer.currentTick);
+        }
+        // Entropy end
         this.setPos(d0, d1, d2);
     }
 
