From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Mon, 29 Jul 2024 15:18:08 +0200
Subject: [PATCH] packet limiter


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 6ea6ddc04d9ca5fc4403edd9c741d0b2514cd348..d4cae2980508490819d3b9542b020b9f1a707215 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -28,6 +28,7 @@ import io.netty.handler.timeout.TimeoutException;
 import io.netty.util.AttributeKey;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Queue;
 import java.util.UUID;
@@ -50,6 +51,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.item.FallingBlockEntity;
 import net.minecraft.world.entity.item.PrimedTnt;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.block.Blocks;
 import org.apache.commons.lang3.Validate;
 import org.bukkit.Bukkit;
@@ -58,6 +60,7 @@ import org.bukkit.entity.FallingBlock;
 import org.slf4j.Logger;
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
+import xyz.fragmentmc.entropy.EntropyConfig;
 import xyz.fragmentmc.entropy.EntropyUserManager;
 import xyz.fragmentmc.entropy.connection.IPMap;
 
@@ -393,44 +396,77 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     public void send(Packet<?> packet, @Nullable PacketSendListener callbacks) {
         // Entropy Start - fps
-        if (getPlayer() != null) {
+        Player player = getPlayer();
+        if (player != null) {
             if (packet instanceof ClientboundMoveEntityPacket p) {
-                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(getPlayer());
-                Entity entity = p.getEntity(getPlayer().level);
+                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(player);
+                Entity entity = p.getEntity(player.level);
                 boolean delete = false;
-                if (entity instanceof PrimedTnt && (data.TntHidden || data.FlashHidden)) {
+                if (entity.isTnt && (data.TntHidden || data.FlashHidden)) {
                     if(entity.tickCount > 1) return;
                     delete = true;
                 }
-                else if(entity.tickCount < 1) {
+                else if(entity.tickCount < 1 && entity.isFallingBlock) {
                     String name = entity.getName().getString();
-                    if (name.contains("Sand") && data.SandHidden) delete = true;
-                    else if (name.contains("Anvil") && data.AnvilHidden) delete = true;
-                    else if (name.contains("Concrete Powder") && data.ConcreteHidden) delete = true;
+                    if (data.SandHidden && name.contains("Sand")) delete = true;
+                    else if (data.AnvilHidden && name.contains("Anvil")) delete = true;
+                    else if (data.ConcreteHidden && name.contains("Concrete Powder")) delete = true;
                 }
                 if (delete) {
-                    send(new ClientboundRemoveEntitiesPacket(entity.getId()));
+                    this.sendPacket(new ClientboundRemoveEntitiesPacket(entity.getId()), null, null);
                     return;
                 }
             }
             else if (packet instanceof ClientboundExplodePacket) {
-                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(getPlayer());
+                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(player);
                 if (data.ExplosionHidden) return;
             }
             else if (packet instanceof ClientboundBlockEventPacket p) {
-                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(getPlayer());
+                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(player);
                 if ((p.getBlock() == Blocks.PISTON || p.getBlock() == Blocks.STICKY_PISTON) && data.PistonHidden) return;
             }
             else if (packet instanceof ClientboundSetEntityDataPacket p) {
-                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(getPlayer());
-                if (getPlayer().level.getEntity(p.id()) instanceof PrimedTnt t && data.FlashHidden) {
+                EntropyUserManager.FpsData data = EntropyUserManager.Instance.getFpsData(player);
+                Entity entity = player.level.getEntity(p.id());
+                if (entity instanceof PrimedTnt t && (data.FlashHidden || t.minimal)) {
                     // https://pokechu22.github.io/Burger/1.20.1.html
-                    ClientboundAddEntityPacket ep = new ClientboundAddEntityPacket(t.getId(), UUID.randomUUID(), t.getX(), t.getY(), t.getZ(), 0, 0, EntityType.FALLING_BLOCK, 2094, t.getDeltaMovement(), 0);
-                    send(ep);
+                    this.sendPacket(new ClientboundRemoveEntitiesPacket(t.getId()), null, null);
+                    t.minimal = true;
+                    if (t.getFuse() > 2) {
+                        ClientboundAddEntityPacket ep = new ClientboundAddEntityPacket(t.getId(), UUID.randomUUID(), t.getX(), t.getY(), t.getZ(), 0, 0, EntityType.FALLING_BLOCK, 2094, t.getDeltaMovement(), 0);
+                        this.sendPacket(ep, null, null);
+                    }
+                    return;
+                }
+            }
+            // Entropy End
+
+            // Entropy Start - packet limiter
+            if (player.tickCount % 5 != 0) { // limit these packets to every 5 ticks
+                if (packet instanceof ClientboundMoveEntityPacket p) {
+                    Entity entity = p.getEntity(player.level);
+                    if (entity == null) return;
+                    else if (entity.isTnt && EntropyConfig.packetLimiterTntMove) return;
+                    else if (entity.isFallingBlock && EntropyConfig.packetLimiterFallingBlockMove) return;
+                }
+                else if (packet instanceof ClientboundSetEntityMotionPacket p) {
+                    Entity entity = player.level.getEntity(p.getId());
+                    if (entity == null) return;
+                    else if (entity.isTnt && EntropyConfig.packetLimiterTntVelocity) return;
+                    else if (entity.isFallingBlock && EntropyConfig.packetLimiterFallingBlockVelocity) return;
+                }
+                else if (packet instanceof ClientboundSetEntityDataPacket p) {
+                    Entity entity = player.level.getEntity(p.id());
+                    if (entity == null) return;
+                    else if (entity.isTnt && EntropyConfig.packetLimiterTntData) return;
+                    else if (entity.isFallingBlock && EntropyConfig.packetLimiterFallingBlockData) return;
+                }
+                else if (packet instanceof ClientboundBlockEventPacket p) {
+                    if ((p.getBlock() == Blocks.PISTON || p.getBlock() == Blocks.STICKY_PISTON) && EntropyConfig.packetLimiterPistonUpdates) return;
                 }
             }
+            // Entropy End
         }
-        // Entropy End
 
         // Paper start - handle oversized packets better
         boolean connected = this.isConnected();
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 8dbf7c5d9e8aa5704c330680513d427496617c02..a56420d7b8c4d180137195deb2533741766d79b8 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -175,6 +175,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
     // Entropy end
 
+    // Entropy start - tick limiter
+    public boolean isTnt = false;
+    public boolean isFallingBlock = false;
+    // Entropy end
+
     // CraftBukkit start
     private static final int CURRENT_LEVEL = 2;
     public boolean preserveMotion = true; // Paper - keep initial motion on first setPositionRotation
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index f0dce3f415b8a8c14c43fac0f63c51426edcb4da..5ece9a9f816a154e0d03b97d9e2f92eeb6a2efff 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -78,6 +78,7 @@ public class FallingBlockEntity extends Entity {
 
     public FallingBlockEntity(EntityType<? extends FallingBlockEntity> type, Level world) {
         super(type, world);
+        isFallingBlock = true;
         this.blockState = Blocks.SAND.defaultBlockState();
         this.dropItem = true;
         this.fallDamageMax = 40;
@@ -93,6 +94,7 @@ public class FallingBlockEntity extends Entity {
         this.yo = y;
         this.zo = z;
         this.setStartPos(this.blockPosition());
+        isFallingBlock = true;
 
         // Entropy Start - entity origin fixes
         this.setOrigin(new Location(world.getWorld(), x, y, z));
diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 81d74571c9414f37b4f41b8a0eee375682eafa92..944cbfe59e1c7bdd05b2373f3a27eb7692e8983a 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -32,6 +32,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     public boolean canCombine = false;
     public boolean blockDamageSkipped;
     public boolean destroyedBlock;
+    public boolean minimal = false;
     // Entropy end
 
     private static final EntityDataAccessor<Integer> DATA_FUSE_ID = SynchedEntityData.defineId(PrimedTnt.class, EntityDataSerializers.INT);
@@ -43,6 +44,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
 
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level world) {
         super(type, world);
+        isTnt = true;
         this.blocksBuilding = true;
     }
 
@@ -63,6 +65,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
 
         // Entropy Start - entity origin fixes
         this.setOrigin(new Location(world.getWorld(), x, y, z));
+        isTnt = true;
     }
 
     @Override
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index e73c0986cfef4b6df023d60acc667efa4997c48d..9ec07ce783d180c0f267e010d28eea7f9920a22b 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -209,10 +209,24 @@ public class EntropyConfig {
     }
 
     public static boolean packetLimiterTntFlow;
+    public static boolean packetLimiterTntMove;
+    public static boolean packetLimiterTntVelocity;
+    public static boolean packetLimiterTntData;
+    public static boolean packetLimiterFallingBlockMove;
+    public static boolean packetLimiterFallingBlockVelocity;
+    public static boolean packetLimiterFallingBlockData;
+    public static boolean packetLimiterPistonUpdates;
     private static void packetLimiter() {
         packetLimiterTntFlow = getBoolean("entropy.packet-limiter.tnt-flow", true, "Update tnt flowing in water less frequently.");
-
-        setComment("entropy.packet-limiter", "Limit packets being sent to the client for performance reasons.");
+        packetLimiterTntMove = getBoolean("entropy.packet-limiter.tnt-move", false, "Update tnt moving less frequently.");
+        packetLimiterTntVelocity = getBoolean("entropy.packet-limiter.tnt-velocity", false, "Update tnt velocity less frequently.");
+        packetLimiterTntData = getBoolean("entropy.packet-limiter.tnt-data", true, "Update tnt data less frequently.");
+        packetLimiterFallingBlockMove = getBoolean("entropy.packet-limiter.falling-block-move", false, "Update falling block movement less frequently.");
+        packetLimiterFallingBlockVelocity = getBoolean("entropy.packet-limiter.falling-block-velocity", false, "Update falling block velocity less frequently.");
+        packetLimiterFallingBlockData = getBoolean("entropy.packet-limiter.falling-block-data", true, "Update falling block data less frequently.");
+        packetLimiterPistonUpdates = getBoolean("entropy.packet-limiter.piston-updates", false, "Update pistons less frequently.");
+
+        setComment("entropy.packet-limiter", "Limit packets being sent to the client for performance reasons. Can cause weird looking behaviour on the client. Not recommended for cannoning servers.");
     }
 
     public static boolean tntLoadsChunks;
