From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Mon, 12 Dec 2022 23:12:25 +0100
Subject: [PATCH] iron golem fall damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index c61318885d7d2602e297c43023c85f898bb446e7..9a660bdc33fd6a22e2f4aa17b529d6438cca03c9 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1821,7 +1821,7 @@ public abstract class LivingEntity extends Entity {
         return SoundEvents.GENERIC_DEATH;
     }
 
-    private SoundEvent getFallDamageSound(int distance) {
+    protected SoundEvent getFallDamageSound(int distance) { // Entropy - enable golem fall damage
         return distance > 4 ? this.getFallSounds().big() : this.getFallSounds().small();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
index e73acfa2f5a4066fa1beee1758082a2fe97a43b3..50bfbaac1d2a92454c696b2c84befcd8831115c6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
@@ -1,12 +1,6 @@
 package net.minecraft.world.entity.animal;
 
 import com.google.common.collect.ImmutableList;
-import java.util.Comparator;
-import java.util.Iterator;
-import java.util.List;
-import java.util.UUID;
-import java.util.stream.Stream;
-import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.BlockParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
@@ -23,20 +17,10 @@ import net.minecraft.util.valueproviders.UniformInt;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.Mob;
-import net.minecraft.world.entity.NeutralMob;
+import net.minecraft.world.entity.*;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
-import net.minecraft.world.entity.ai.goal.GolemRandomStrollInVillageGoal;
-import net.minecraft.world.entity.ai.goal.LookAtPlayerGoal;
-import net.minecraft.world.entity.ai.goal.MeleeAttackGoal;
-import net.minecraft.world.entity.ai.goal.MoveBackToVillageGoal;
-import net.minecraft.world.entity.ai.goal.MoveTowardsTargetGoal;
-import net.minecraft.world.entity.ai.goal.OfferFlowerGoal;
-import net.minecraft.world.entity.ai.goal.RandomLookAroundGoal;
+import net.minecraft.world.entity.ai.goal.*;
 import net.minecraft.world.entity.ai.goal.target.DefendVillageTargetGoal;
 import net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal;
 import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
@@ -52,6 +36,14 @@ import net.minecraft.world.level.NaturalSpawner;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.Vec3;
+import xyz.fragmentmc.entropy.EntropyConfig;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Stream;
 
 public class IronGolem extends AbstractGolem implements NeutralMob {
 
@@ -69,6 +61,33 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         this.maxUpStep = 1.0F;
     }
 
+    // Entropy start - enable golem fall damage
+    // most of this is just a copy of normal fall damage code...
+    @Override
+    public boolean causeFallDamage(float fallDistance, float damageMultiplier, DamageSource damageSource) {
+        if (EntropyConfig.golemEnableFallDamage) {
+            if (this.isVehicle()) {
+                Iterator iterator = this.getPassengers().iterator();
+                while (iterator.hasNext()) {
+                    Entity entity = (Entity) iterator.next();
+                    entity.causeFallDamage(fallDistance, damageMultiplier, damageSource);
+                }
+            }
+            int i = this.calculateFallDamage(fallDistance, damageMultiplier);
+
+            if (i > 0) {
+                if (!this.hurt(damageSource, (float) i)) {
+                    return true;
+                }
+                this.playSound(this.getFallDamageSound(i), 1.0F, 1.0F);
+                this.playBlockFallSound();
+                return true;
+            }
+        }
+        return false;
+    }
+    // Entropy end
+
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(1, new MeleeAttackGoal(this, 1.0D, true));
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index e2b91c7ebde726ee6ad0e7bb2df532b6713ea0e6..bc7ea1b768236150363d7d5a1a3774aaf06f0a5a 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -275,7 +275,7 @@ public class EntropyConfig {
     public static int pearlRangeLimitChunks;
 
     public static boolean playerIgnoreTntDamage;
-
+    public static boolean golemEnableFallDamage;
     public static double spawnerEntityUpwardsVelocity;
     public static boolean spawnerOverrideEntityHealth;
     public static float spawnerEntityHealth;
@@ -355,6 +355,8 @@ public class EntropyConfig {
 
         playerIgnoreTntDamage= getBoolean("entropy.factions.player-ignore-explosion-damage", true,
                 "Players will not take damage from explosions caused by tnt.");
+        golemEnableFallDamage= getBoolean("entropy.factions.golem-enable-fall-damage", false,
+                "Enable fall damage for iron golems.");
         spawnerEntityUpwardsVelocity= getDouble("entropy.factions.spawner.entity-upwards-velocity", 0,
                 "Initial Y velocity for entities spawned by spawners, can have decimals.");
         spawnerOverrideEntityHealth= getBoolean("entropy.factions.spawner.entity-override-health", false,
