From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Mon, 12 Dec 2022 23:41:04 +0100
Subject: [PATCH] disable entity death animation


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 9a660bdc33fd6a22e2f4aa17b529d6438cca03c9..6148583cbaed25107928b6c24516dbfc32c4cb38 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1655,8 +1655,14 @@ public abstract class LivingEntity extends Entity {
             }
 
             if (this.dead) { // Paper
-            this.level.broadcastEntityEvent(this, (byte) 3);
-            this.setPose(Pose.DYING);
+                // Entropy start - disable entity death animation
+                if(EntropyConfig.disableEntityDeathAnimation && this.level instanceof ServerLevel server) {
+                    server.getChunkSource().broadcastAndSend(this, new ClientboundRemoveEntitiesPacket(this.getId()));
+                } else {
+                    this.level.broadcastEntityEvent(this, (byte) 3);
+                }
+                // Entropy end
+                this.setPose(Pose.DYING);
             } // Paper
         }
     }
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index bc7ea1b768236150363d7d5a1a3774aaf06f0a5a..2b7cad19edbe33eac336a4dc48059539ba8ce446 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -276,6 +276,7 @@ public class EntropyConfig {
 
     public static boolean playerIgnoreTntDamage;
     public static boolean golemEnableFallDamage;
+    public static boolean disableEntityDeathAnimation;
     public static double spawnerEntityUpwardsVelocity;
     public static boolean spawnerOverrideEntityHealth;
     public static float spawnerEntityHealth;
@@ -357,6 +358,8 @@ public class EntropyConfig {
                 "Players will not take damage from explosions caused by tnt.");
         golemEnableFallDamage= getBoolean("entropy.factions.golem-enable-fall-damage", false,
                 "Enable fall damage for iron golems.");
+        disableEntityDeathAnimation= getBoolean("entropy.factions.disable-entity-death-animation", false,
+                "Disable the annoying death animation for all entities,");
         spawnerEntityUpwardsVelocity= getDouble("entropy.factions.spawner.entity-upwards-velocity", 0,
                 "Initial Y velocity for entities spawned by spawners, can have decimals.");
         spawnerOverrideEntityHealth= getBoolean("entropy.factions.spawner.entity-override-health", false,
