From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Tue, 13 Dec 2022 19:02:11 +0100
Subject: [PATCH] remove blockstates from api


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 5006315564a07607aaaa27a00cf2a6516983bd6e..26d89a96fdeb20192116709813f967c71bee4692 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -40,8 +40,6 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import org.bukkit.Location;
-import org.bukkit.craftbukkit.block.CraftBlockState;
-import org.bukkit.craftbukkit.block.CraftBlockStates;
 import org.bukkit.craftbukkit.entity.CraftEntity;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.block.BlockExplodeEvent;
@@ -195,7 +193,7 @@ public class Explosion {
         boolean doBlockBreakCheck = true;
         if(EntropyConfig.explosionWaterCheck && this.source != null) {
             BlockState blockState = this.level.getBlockState(this.source.blockPosition);
-            doBlockBreakCheck=!(EventUtil.sendCancellable(level, new ExplodeLavaEvent(EntropyConfig.explodeLava, craftSource, location, null)).isCancelled() ? blockState.getMaterial().equals(Material.WATER) : blockState.getMaterial().isLiquid());
+            doBlockBreakCheck=!(EventUtil.sendCancellable(level, new ExplodeLavaEvent(EntropyConfig.explodeLava, craftSource, location)).isCancelled() ? blockState.getMaterial().equals(Material.WATER) : blockState.getMaterial().isLiquid());
         }
         int projectionSize = EntropyConfig.reduceRays ? EntropyConfig.rayProjectionSize : 15; // Entropy - reduce rays
         if(doBlockBreakCheck) // Entropy end
@@ -255,11 +253,10 @@ public class Explosion {
                             // Entropy end
 
                             // Entropy start - explode waterlogged blocks and lava
-                            CraftBlockState craftBlockState = CraftBlockStates.getBlockState(level, blockposition);
-                            if (!iblockdata.getFluidState().isEmpty() && !iblockdata.getMaterial().isLiquid() && !EventUtil.sendCancellable(level, new ExplodeWaterloggedEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location, craftBlockState)).isCancelled()) {
+                            if (!iblockdata.getFluidState().isEmpty() && !iblockdata.getMaterial().isLiquid() && !EventUtil.sendCancellable(level, new ExplodeWaterloggedEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location)).isCancelled()) {
                                 blockDamageSet.add(blockposition);
                             }
-                            if (iblockdata.getMaterial() == Material.LAVA  && !EventUtil.sendCancellable(level, new ExplodeLavaEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location, craftBlockState)).isCancelled()) {
+                            if (iblockdata.getMaterial() == Material.LAVA  && !EventUtil.sendCancellable(level, new ExplodeLavaEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location)).isCancelled()) {
                                 blockDamageSet.add(blockposition);
                             }
                             // Entropy end
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
index a8ab1d3ee81664193be39d2735d6495136e0e310..80d54a5c8206ce698ff02b71e55865cd0900b08d 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
@@ -1,9 +1,6 @@
 package org.bukkit.craftbukkit.block;
 
 import com.google.common.base.Preconditions;
-import java.lang.ref.WeakReference;
-import java.util.List;
-import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.LevelAccessor;
 import org.bukkit.Chunk;
@@ -21,6 +18,10 @@ import org.bukkit.material.MaterialData;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
 
+import javax.annotation.Nullable;
+import java.lang.ref.WeakReference;
+import java.util.List;
+
 public class CraftBlockState implements BlockState {
 
     protected final CraftWorld world;
@@ -42,7 +43,7 @@ public class CraftBlockState implements BlockState {
     }
 
     // world can be null for non-placed BlockStates.
-    protected CraftBlockState(@Nullable World world, BlockPos blockPosition, net.minecraft.world.level.block.state.BlockState blockData) {
+    public CraftBlockState(@Nullable World world, BlockPos blockPosition, net.minecraft.world.level.block.state.BlockState blockData) { // Entropy - optimise durability event
         this.world = (CraftWorld) world;
         this.position = blockPosition;
         this.data = blockData;
diff --git a/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java b/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
index 871c7b1302d9ad13209082714ee17fa2a0c75a54..3893503468b76cf3e2b10fad2fee52cd2396ef93 100644
--- a/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
+++ b/src/main/java/xyz/fragmentmc/entropy/UnbreakableDurability.java
@@ -8,8 +8,8 @@ import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.Location;
 import org.bukkit.Material;
-import org.bukkit.craftbukkit.block.CraftBlockStates;
 import xyz.fragmentmc.entropy.event.AssignDurabilityEvent;
 
 import java.util.HashMap;
@@ -89,7 +89,7 @@ public class UnbreakableDurability {
                     return false;
                 }
             }
-            AssignDurabilityEvent event = new AssignDurabilityEvent(durability, CraftBlockStates.getBlockState(level, b));
+            AssignDurabilityEvent event = new AssignDurabilityEvent(durability, bs.getBukkitMaterial(), new Location(level.getWorld(), b.getX(), b.getY(), b.getZ()));
             level.getCraftServer().getPluginManager().callEvent(event);
             unbreakableDurability.put(b.asLong(), event.getDurability());
         }
