From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Wed, 14 Dec 2022 19:45:17 +0100
Subject: [PATCH] fix combined explosions api event


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 26d89a96fdeb20192116709813f967c71bee4692..440e8f96a48734799c08059afc8b49b69f608b83 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -167,7 +167,7 @@ public class Explosion {
 
         // Entropy start - combined explosions
         int merged = 1;
-        if (this.source instanceof PrimedTnt && ((PrimedTnt) this.source).canCombine && !EventUtil.sendCancellable(level, new CombinedExplosionEvent(true, craftSource, location)).isCancelled()) {
+        if (this.source instanceof PrimedTnt && ((PrimedTnt) this.source).canCombine && !EventUtil.sendCancellable(level, new CombinedExplosionEvent(false, craftSource, location)).isCancelled()) {
             merged = ((PrimedTnt) this.source).count;
         }
         // Entropy end
