From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Wed, 11 Jan 2023 15:05:44 +0100
Subject: [PATCH] remove incoming chat message verification and disable kicks


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1a34719de7ab303250a38c2cfbeb2ed9cc0f5ac5..88951faac4682b638d8041397fb4f947f8fefe82 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2086,6 +2086,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             return;
         }
         // CraftBukkit end
+
         if (ServerGamePacketListenerImpl.isChatMessageIllegal(packet.message())) {
             this.server.scheduleOnMain(() -> { // Paper - push to main for event firing
             this.disconnect(Component.translatable("multiplayer.disconnect.illegal_characters"), org.bukkit.event.player.PlayerKickEvent.Cause.ILLEGAL_CHARACTERS); // Paper - add cause
@@ -2093,9 +2094,9 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         } else {
             if (this.tryHandleChat(packet.message(), packet.timeStamp(), packet.lastSeenMessages())) {
                 // this.server.submit(() -> { // CraftBukkit - async chat
-                    PlayerChatMessage playerchatmessage = this.getSignedMessage(packet);
+                    PlayerChatMessage playerchatmessage = EntropyConfig.disableChatMessageVerification ? PlayerChatMessage.unsigned(new MessageSigner(UUID.randomUUID(), Instant.now(), packet.salt()), new ChatMessageContent(packet.message())) : this.getSignedMessage(packet); // Entropy - disable chat message verification
 
-                    if (this.verifyChatMessage(playerchatmessage)) {
+                    if (EntropyConfig.disableChatMessageVerification || this.verifyChatMessage(playerchatmessage)) { // Entropy - disable chat message verification
                         this.chatMessageChain.append(() -> {
                             CompletableFuture<FilteredText> completablefuture = this.filterTextPacket(playerchatmessage.signedContent().plain());
                             CompletableFuture<PlayerChatMessage> completablefuture1 = this.server.getChatDecorator().decorate(this.player, null, playerchatmessage); // Paper
@@ -2612,8 +2613,12 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     }
 
     private void handleValidationFailure(Set<LastSeenMessagesValidator.ErrorCondition> reasons) {
-        ServerGamePacketListenerImpl.LOGGER.warn("Failed to validate message from {}, reasons: {}", this.player.getName().getString(), reasons.stream().map(LastSeenMessagesValidator.ErrorCondition::message).collect(Collectors.joining(",")));
-        this.disconnect(Component.translatable("multiplayer.disconnect.chat_validation_failed"), org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event cause
+        // Entropy start - disable chat message verification
+        if(!EntropyConfig.disableChatMessageVerification) {
+            ServerGamePacketListenerImpl.LOGGER.warn("Failed to validate message from {}, reasons: {}", this.player.getName().getString(), reasons.stream().map(LastSeenMessagesValidator.ErrorCondition::message).collect(Collectors.joining(",")));
+            this.disconnect(Component.translatable("multiplayer.disconnect.chat_validation_failed"), org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event cause
+        }
+        // Entropy end
     }
 
     @Override
