From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Wed, 11 Jan 2023 18:42:45 +0100
Subject: [PATCH] no boat collisions


diff --git a/src/main/java/io/papermc/paper/util/CollisionUtil.java b/src/main/java/io/papermc/paper/util/CollisionUtil.java
index de780f8b49cb087e65114dbcb6cc73a49cef087a..83685125788fa100df133576cc29d1d946ec306e 100644
--- a/src/main/java/io/papermc/paper/util/CollisionUtil.java
+++ b/src/main/java/io/papermc/paper/util/CollisionUtil.java
@@ -6,6 +6,7 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.*;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.vehicle.Boat;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.level.CollisionGetter;
 import net.minecraft.world.level.EntityGetter;
@@ -20,6 +21,7 @@ import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.*;
+import xyz.fragmentmc.entropy.EntropyConfig;
 
 import java.util.List;
 import java.util.function.BiPredicate;
@@ -831,7 +833,7 @@ public final class CollisionUtil {
 
             for (int i = 0, len = entities.size(); i < len; ++i) {
                 final Entity otherEntity = entities.get(i);
-
+                if (EntropyConfig.noBoatCollisions && (entity instanceof PrimedTnt || entity instanceof FallingBlockEntity) && otherEntity instanceof Boat) continue;
                 if ((entity == null && otherEntity.canBeCollidedWith()) || (entity != null && entity.canCollideWith(otherEntity))) {
                     if (checkOnly) {
                         return true;
diff --git a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
index 2b7cad19edbe33eac336a4dc48059539ba8ce446..c1dc3c77a291cca7493772aa8b55ddedcc854b99 100644
--- a/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
+++ b/src/main/java/xyz/fragmentmc/entropy/EntropyConfig.java
@@ -244,6 +244,7 @@ public class EntropyConfig {
     public static boolean entityEyeHeightParity;
     public static boolean explodeWaterloggedBlocks;
     public static boolean explodeLava;
+    public static boolean noBoatCollisions;
 
     public static boolean mergeEntities;
     public static boolean combinedExplosions;
@@ -294,6 +295,8 @@ public class EntropyConfig {
         explodeLava= getBoolean("entropy.patches.explode-lava", false,
                 "Lava will be destroyed by explosions.",
                 "Use this for non-regen servers.");
+        noBoatCollisions = getBoolean("entropy.patches.no-boat-collisions", false,
+                "Disable tnt and falling blocks colliding with boats.");
         setComment("entropy.patches", "Changes to game mechanics related to cannoning");
 
         mergeEntities= getBoolean("entropy.optimizations.merge-entities", true,
