From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <dusek.s03@htlwienwest.at>
Date: Sat, 14 Jan 2023 18:54:24 +0100
Subject: [PATCH] non-watered combined explosions


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index 90fd80f6f69fc9dee51b0d5eb432021bb16e3349..de8b2a9d829e6ec32dac2778b248a339ad05b939 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -32,6 +32,8 @@ public class PrimedTnt extends Entity {
 
     public boolean canCombine = false; // Entropy - combined explosions
     public int count = 1; // Entropy - merge cannoning entities
+    public boolean blockDamageSkipped;
+    public boolean destroyedBlock;
 
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level world) {
         super(type, world);
@@ -73,22 +75,29 @@ public class PrimedTnt extends Entity {
     @Override
     public void tick() {
         // Entropy start - merge cannoning entities
-        if (this.getFuse() - 1 <= 0 && this.count > 1) {
-            PrimedTnt vtnt = this.clone();
-            vtnt.setFuse(100);
-            vtnt.tick();
-
-            if (EntropyConfig.mergeEntities &&
-                    vtnt.level.getBlockState(vtnt.blockPosition()).getMaterial().isLiquid() &&
-                    ((this.getDeltaMovement().x == 0 && this.getDeltaMovement().y == 0 && this.getDeltaMovement().z == 0) || this.distanceTo(vtnt) > 8)) {
-                this.canCombine = true;
-            } else {
-                for (int x = 0; x < this.count -1; x++) {
+        if (this.getFuse() - 1 <= 0 && !this.canCombine) {
+            int tc = this.count;
+                for (int x = 0; x < tc -1; x++) {
                     PrimedTnt merged = this.clone();
                     merged.tick();
+                    this.count--;
+                    if (EntropyConfig.mergeEntities &&
+                            (
+                                (
+                                    EntropyConfig.staticRayStrenght && !merged.destroyedBlock
+                                ) ||
+                                merged.blockDamageSkipped
+                            ) &&
+                            (
+                                this.distanceTo(merged) == 0 ||
+                                this.distanceTo(merged) > 8
+                            )
+                    ) {
+                        System.out.println(count);
+                        this.canCombine = true;
+                        break;
+                    }
                 }
-            }
-            vtnt.discard();
         }
         // Entropy end
         if (level.spigotConfig.maxTntTicksPerTick > 0 && ++level.spigotConfig.currentPrimedTnt > level.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 440e8f96a48734799c08059afc8b49b69f608b83..876de6a8683d2c7b23d1385002be03151d7ae17a 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -270,7 +270,7 @@ public class Explosion {
                                     unbreakables.add(blockposition);
                             }
                             else // Entropy end
-                            if (f > 0.0F && this.damageCalculator.shouldBlockExplode(this, this.level, blockposition, iblockdata, f)) {
+                            if (f > 0.0F && this.damageCalculator.shouldBlockExplode(this, this.level, blockposition, iblockdata, f) && !iblockdata.isAir()) { //Entropy - non-watered combined explosions
                                 blockDamageSet.add(blockposition); // Entropy - replace hash set
                                 // Paper start - prevent headless pistons from forming
                                 if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowHeadlessPistons && iblockdata.getBlock() == Blocks.MOVING_PISTON) {
@@ -291,6 +291,13 @@ public class Explosion {
                 }
             }
         }
+        // Entropy start - non-watered combined explosions
+        else {
+            if (this.source instanceof PrimedTnt t) t.blockDamageSkipped = true;
+        }
+        if(this.source instanceof PrimedTnt t && (unbreakables.size() > 0 || blockDamageSet.size() > 0)) t.destroyedBlock =  true;
+        // Entropy end
+
         // Entropy start - unbreakable breaker
         for(BlockPos b : unbreakables) {
             if (level.unbreakableDurability.reduceDurability(b)) blockDamageSet.add(b);
