From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Thu, 9 Feb 2023 21:38:37 +0100
Subject: [PATCH] fix explode lava


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 3e8a1718dbe3b036d66fb59c86b127c263ad00d4..72d3f2c1a8e0853545c2f6a863d1613830490477 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -256,7 +256,7 @@ public class Explosion {
                             if (!iblockdata.getFluidState().isEmpty() && !iblockdata.getMaterial().isLiquid() && !EventUtil.sendCancellable(level, new ExplodeWaterloggedEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location)).isCancelled()) {
                                 blockDamageSet.add(blockposition);
                             }
-                            if (iblockdata.getMaterial() == Material.LAVA  && !EventUtil.sendCancellable(level, new ExplodeLavaEvent(!EntropyConfig.explodeWaterloggedBlocks, craftSource, location)).isCancelled()) {
+                            if (iblockdata.getMaterial() == Material.LAVA  && !EventUtil.sendCancellable(level, new ExplodeLavaEvent(!EntropyConfig.explodeLava, craftSource, location)).isCancelled()) {
                                 blockDamageSet.add(blockposition);
                             }
                             // Entropy end
