From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stan <77831846+stanuwu@users.noreply.github.com>
Date: Thu, 9 Feb 2023 21:52:01 +0100
Subject: [PATCH] fix submerged check with explode lava


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 72d3f2c1a8e0853545c2f6a863d1613830490477..dd43f4b20b909243353002af069d67e26e7b4cc3 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -193,7 +193,7 @@ public class Explosion {
         boolean doBlockBreakCheck = true;
         if(EntropyConfig.explosionWaterCheck && this.source != null) {
             BlockState blockState = this.level.blockStateCache.getCached(level, this.source.blockPosition());
-            doBlockBreakCheck=!((!EventUtil.sendCancellable(level, new ExplodeLavaEvent(EntropyConfig.explodeLava, craftSource, location)).isCancelled()) ? blockState.getMaterial().equals(Material.WATER) : blockState.getMaterial().isLiquid());
+            doBlockBreakCheck=!((!EventUtil.sendCancellable(level, new ExplodeLavaEvent(!EntropyConfig.explodeLava, craftSource, location)).isCancelled()) ? blockState.getMaterial().equals(Material.WATER) : blockState.getMaterial().isLiquid());
         }
         int projectionSize = EntropyConfig.reduceRays ? EntropyConfig.rayProjectionSize : 15; // Entropy - reduce rays
         if(doBlockBreakCheck) // Entropy end
